

<!DOCTYPE html>
<html class="writer-html5" lang="en">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>spac package &mdash; spac  documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="_static/custom.css?v=b3f5c315" />

  
      <script src="_static/jquery.js?v=5d32c60e"></script>
      <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js?v=b3ba4146"></script>
      <script src="_static/doctools.js?v=888ff710"></script>
      <script src="_static/sphinx_highlight.js?v=4825356b"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            spac
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="modules/index.html">spac modules</a><ul>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.data_utils.add_pin_color_rules.html">add_pin_color_rules</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.data_utils.add_rescaled_features.html">add_rescaled_features</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.data_utils.append_annotation.html">append_annotation</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.data_utils.bin2cat.html">bin2cat</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.data_utils.calculate_centroid.html">calculate_centroid</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.data_utils.combine_annotations.html">combine_annotations</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.data_utils.combine_dfs.html">combine_dfs</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.data_utils.concatinate_regions.html">concatinate_regions</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.data_utils.downsample_cells.html">downsample_cells</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.data_utils.ingest_cells.html">ingest_cells</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.data_utils.load_csv_files.html">load_csv_files</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.data_utils.rescale_features.html">rescale_features</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.data_utils.html">spac.data_utils module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.data_utils.append_annotation.html">append_annotation</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.data_utils.ingest_cells.html">ingest_cells</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.data_utils.concatinate_regions.html">concatinate_regions</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.data_utils.rescale_features.html">rescale_features</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.data_utils.add_rescaled_features.html">add_rescaled_features</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.data_utils.subtract_min_per_region.html">subtract_min_per_region</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.data_utils.subtract_min_quantile.html">subtract_min_quantile</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.data_utils.load_csv_files.html">load_csv_files</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.data_utils.select_values.html">select_values</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.data_utils._select_values_dataframe.html">_select_values_dataframe</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.data_utils._select_values_anndata.html">_select_values_anndata</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.data_utils.downsample_cells.html">downsample_cells</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.data_utils.calculate_centroid.html">calculate_centroid</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.data_utils.bin2cat.html">bin2cat</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.data_utils.combine_dfs.html">combine_dfs</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.data_utils.add_pin_color_rules.html">add_pin_color_rules</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.data_utils.combine_annotations.html">combine_annotations</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.data_utils.summarize_dataframe.html">summarize_dataframe</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.data_utils._select_values_anndata.html">_select_values_anndata</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.data_utils._select_values_dataframe.html">_select_values_dataframe</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.data_utils.select_values.html">select_values</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.data_utils.subtract_min_per_region.html">subtract_min_per_region</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.data_utils.subtract_min_quantile.html">subtract_min_quantile</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.data_utils.summarize_dataframe.html">summarize_dataframe</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.phenotyping.apply_phenotypes.html">apply_phenotypes</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.phenotyping.assign_manual_phenotypes.html">assign_manual_phenotypes</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.phenotyping.combine_phenotypes.html">combine_phenotypes</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.phenotyping.decode_phenotype.html">decode_phenotype</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.phenotyping.generate_phenotypes_dict.html">generate_phenotypes_dict</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.phenotyping.is_binary_0_1.html">is_binary_0_1</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.phenotyping.html">spac.phenotyping module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.phenotyping.is_binary_0_1.html">is_binary_0_1</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.phenotyping.decode_phenotype.html">decode_phenotype</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.phenotyping.generate_phenotypes_dict.html">generate_phenotypes_dict</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.phenotyping.apply_phenotypes.html">apply_phenotypes</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.phenotyping.combine_phenotypes.html">combine_phenotypes</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.phenotyping.assign_manual_phenotypes.html">assign_manual_phenotypes</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.spatial_analysis.calculate_nearest_neighbor.html">calculate_nearest_neighbor</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.spatial_analysis._neighborhood_profile_core.html">_neighborhood_profile_core</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.spatial_analysis.neighborhood_profile.html">neighborhood_profile</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.spatial_analysis.ripley_l.html">ripley_l</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.spatial_analysis.html">spac.spatial_analysis module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.spatial_analysis.spatial_interaction.html">spatial_interaction</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.spatial_analysis.ripley_l.html">ripley_l</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.spatial_analysis.neighborhood_profile.html">neighborhood_profile</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.spatial_analysis._neighborhood_profile_core.html">_neighborhood_profile_core</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.spatial_analysis.calculate_nearest_neighbor.html">calculate_nearest_neighbor</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.spatial_analysis.spatial_interaction.html">spatial_interaction</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.transformations.apply_per_batch.html">apply_per_batch</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.transformations.arcsinh_transformation_core.html">arcsinh_transformation_core</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.transformations.arcsinh_transformation.html">arcsinh_transformation</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.transformations.batch_normalize.html">batch_normalize</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.transformations.get_cluster_info.html">get_cluster_info</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.transformations.knn_clustering.html">knn_clustering</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.transformations.normalize_features_core.html">normalize_features_core</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.transformations.normalize_features.html">normalize_features</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.transformations.phenograph_clustering.html">phenograph_clustering</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.transformations.rename_annotations.html">rename_annotations</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.transformations.html">spac.transformations module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.transformations.phenograph_clustering.html">phenograph_clustering</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.transformations.knn_clustering.html">knn_clustering</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.transformations.get_cluster_info.html">get_cluster_info</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.transformations.tsne.html">tsne</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.transformations.run_umap.html">run_umap</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.transformations._validate_transformation_inputs.html">_validate_transformation_inputs</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.transformations._select_input_features.html">_select_input_features</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.transformations.batch_normalize.html">batch_normalize</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.transformations.rename_annotations.html">rename_annotations</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.transformations.normalize_features.html">normalize_features</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.transformations.normalize_features_core.html">normalize_features_core</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.transformations.arcsinh_transformation.html">arcsinh_transformation</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.transformations.arcsinh_transformation_core.html">arcsinh_transformation_core</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.transformations.z_score_normalization.html">z_score_normalization</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.transformations.apply_per_batch.html">apply_per_batch</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.transformations.run_utag_clustering.html">run_utag_clustering</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.transformations.run_umap.html">run_umap</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.transformations.run_utag_clustering.html">run_utag_clustering</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.transformations._select_input_features.html">_select_input_features</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.transformations.tsne.html">tsne</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.transformations._validate_transformation_inputs.html">_validate_transformation_inputs</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.transformations.z_score_normalization.html">z_score_normalization</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.utag_functions.add_probabilities_to_centroid.html">add_probabilities_to_centroid</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.utag_functions.custom_message_passing.html">custom_message_passing</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.utag_functions.low_variance_filter.html">low_variance_filter</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.utag_functions._parallel_message_pass.html">_parallel_message_pass</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.utag_functions.html">spac.utag_functions module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.utag_functions.z_score.html">z_score</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.utag_functions.sparse_matrix_dstack.html">sparse_matrix_dstack</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.utag_functions._parallel_message_pass.html">_parallel_message_pass</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.utag_functions.custom_message_passing.html">custom_message_passing</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.utag_functions.low_variance_filter.html">low_variance_filter</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.utag_functions.add_probabilities_to_centroid.html">add_probabilities_to_centroid</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.utag_functions.utag.html">utag</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.utag_functions.sparse_matrix_dstack.html">sparse_matrix_dstack</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.utag_functions.utag.html">utag</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.utag_functions.z_score.html">z_score</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.utils.annotation_category_relations.html">annotation_category_relations</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.utils.check_annotation.html">check_annotation</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.utils.check_column_name.html">check_column_name</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.utils.check_distances.html">check_distances</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.utils.check_feature.html">check_feature</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.utils.check_label.html">check_label</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.utils.check_list_in_list.html">check_list_in_list</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.utils.check_table.html">check_table</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.utils.color_mapping.html">color_mapping</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.utils.compute_boxplot_metrics.html">compute_boxplot_metrics</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.utils.get_defined_color_map.html">get_defined_color_map</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.utils.regex_search_list.html">regex_search_list</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.utils.html">spac.utils module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.utils.regex_search_list.html">regex_search_list</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.utils.check_list_in_list.html">check_list_in_list</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.utils.check_table.html">check_table</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.utils.check_annotation.html">check_annotation</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.utils.check_feature.html">check_feature</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.utils.check_column_name.html">check_column_name</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.utils.check_distances.html">check_distances</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.utils.text_to_others.html">text_to_others</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.utils.annotation_category_relations.html">annotation_category_relations</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.utils.color_mapping.html">color_mapping</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.utils.check_label.html">check_label</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.utils.spell_out_special_characters.html">spell_out_special_characters</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.utils.get_defined_color_map.html">get_defined_color_map</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.utils.compute_boxplot_metrics.html">compute_boxplot_metrics</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.utils.spell_out_special_characters.html">spell_out_special_characters</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.utils.text_to_others.html">text_to_others</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.visualization.boxplot_interactive.html">boxplot_interactive</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.visualization.boxplot.html">boxplot</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.visualization._css_rgb_or_hex_to_hex.html">_css_rgb_or_hex_to_hex</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.visualization.dimensionality_reduction_plot.html">dimensionality_reduction_plot</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.visualization.heatmap.html">heatmap</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.visualization.hierarchical_heatmap.html">hierarchical_heatmap</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.visualization.histogram.html">histogram</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.visualization.interactive_spatial_plot.html">interactive_spatial_plot</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.visualization._ordered_unique_figs.html">_ordered_unique_figs</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.visualization.plot_ripley_l.html">plot_ripley_l</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.visualization._plot_spatial_distance_dispatch.html">_plot_spatial_distance_dispatch</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.visualization._prepare_spatial_distance_data.html">_prepare_spatial_distance_data</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.visualization.present_summary_as_figure.html">present_summary_as_figure</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.visualization.present_summary_as_html.html">present_summary_as_html</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.visualization.relational_heatmap.html">relational_heatmap</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.visualization.html">spac.visualization module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.visualization.visualize_2D_scatter.html">visualize_2D_scatter</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.visualization.dimensionality_reduction_plot.html">dimensionality_reduction_plot</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.visualization.tsne_plot.html">tsne_plot</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.visualization.histogram.html">histogram</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.visualization.heatmap.html">heatmap</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.visualization.hierarchical_heatmap.html">hierarchical_heatmap</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.visualization.threshold_heatmap.html">threshold_heatmap</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.visualization.spatial_plot.html">spatial_plot</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.visualization.boxplot.html">boxplot</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.visualization.boxplot_interactive.html">boxplot_interactive</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.visualization.interactive_spatial_plot.html">interactive_spatial_plot</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.visualization.sankey_plot.html">sankey_plot</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.visualization.relational_heatmap.html">relational_heatmap</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.visualization.plot_ripley_l.html">plot_ripley_l</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.visualization._prepare_spatial_distance_data.html">_prepare_spatial_distance_data</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.visualization._plot_spatial_distance_dispatch.html">_plot_spatial_distance_dispatch</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.visualization._css_rgb_or_hex_to_hex.html">_css_rgb_or_hex_to_hex</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.visualization._ordered_unique_figs.html">_ordered_unique_figs</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.visualization.visualize_nearest_neighbor.html">visualize_nearest_neighbor</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.visualization.present_summary_as_html.html">present_summary_as_html</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/spac.visualization.present_summary_as_figure.html">present_summary_as_figure</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.visualization.sankey_plot.html">sankey_plot</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.visualization.spatial_plot.html">spatial_plot</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.visualization.threshold_heatmap.html">threshold_heatmap</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.visualization.tsne_plot.html">tsne_plot</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.visualization.visualize_2D_scatter.html">visualize_2D_scatter</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/spac.visualization.visualize_nearest_neighbor.html">visualize_nearest_neighbor</a></li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">spac</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content style-external-links">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">spac package</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/spac.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="spac-package">
<h1>spac package<a class="headerlink" href="#spac-package" title="Permalink to this heading"></a></h1>
<p>Version: 0.9.0
Author: Fang Liu, Rui He, and George Zaki</p>
</section>
<section id="module-spac.data_utils">
<span id="submodules-functions"></span><h1>Submodules &amp; Functions<a class="headerlink" href="#module-spac.data_utils" title="Permalink to this heading"></a></h1>
<dl class="py function">
<dt class="sig sig-object py" id="spac.data_utils.add_pin_color_rules">
<span class="sig-prename descclassname"><span class="pre">spac.data_utils.</span></span><span class="sig-name descname"><span class="pre">add_pin_color_rules</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">label_color_dict</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color_map_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'_spac_colors'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">dict</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/spac/data_utils.html#add_pin_color_rules"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.data_utils.add_pin_color_rules" title="Permalink to this definition"></a></dt>
<dd><p>Adds pin color rules to the AnnData object and scans for matching labels.</p>
<p>This function scans unique labels in each adata.obs and column names in all
adata tables, to find the labels defined by the pin color rule.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – The anndata object containing upstream analysis.</p></li>
<li><p><strong>label_color_dict</strong> (<em>dict</em>) – Dictionary of pin color rules with label as key and color as value.</p></li>
<li><p><strong>color_map_name</strong> (<em>str</em>) – The name to use for storing pin color rules in <cite>adata.uns</cite>.</p></li>
<li><p><strong>overwrite</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to overwrite existing pin color rules in <cite>adata.uns</cite> with the
same name, by default True.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>label_matches</strong> (<em>dict</em>) – Dictionary with the matching labels in each
section (obs, var, X, etc.).</p></li>
<li><p><strong>result_str</strong> (<em>str</em>) – Summary string with the matching labels in each
section (obs, var, X, etc.).</p></li>
</ul>
</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ValueError</strong> – If <cite>color_map_name</cite> already exists in <cite>adata.uns</cite>
    and <cite>overwrite</cite> is False.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.data_utils.add_rescaled_features">
<span class="sig-prename descclassname"><span class="pre">spac.data_utils.</span></span><span class="sig-name descname"><span class="pre">add_rescaled_features</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_quantile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_quantile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/data_utils.html#add_rescaled_features"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.data_utils.add_rescaled_features" title="Permalink to this definition"></a></dt>
<dd><p>Clip and rescale the features matrix.</p>
<p>The results will be added into a new layer in the AnnData object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>anndata.AnnData</em>) – The AnnData object.</p></li>
<li><p><strong>min_quantile</strong> (<em>float</em>) – The minimum quantile to rescale to zero.</p></li>
<li><p><strong>max_quantile</strong> (<em>float</em>) – The maximum quantile to rescale to one.</p></li>
<li><p><strong>layer</strong> (<em>str</em>) – The name of the new layer to add to the anndata object.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.data_utils.append_annotation">
<span class="sig-prename descclassname"><span class="pre">spac.data_utils.</span></span><span class="sig-name descname"><span class="pre">append_annotation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="reference internal" href="_modules/spac/data_utils.html#append_annotation"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.data_utils.append_annotation" title="Permalink to this definition"></a></dt>
<dd><p>Append a new annotation with single value to
a Pandas DataFrame based on mapping rules.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>pd.DataFrame</em>) – The input DataFrame to which the new observation will be appended.</p></li>
<li><p><strong>annotation</strong> (<em>dict</em>) – dictionary of string pairs representing
the new annotation and its value.
Each pair should have this format:
&lt;new annotation column name&gt;:&lt;value of the annotation&gt;
The values must be a single string or numeric value.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The DataFrame with the new observation appended.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.data_utils.bin2cat">
<span class="sig-prename descclassname"><span class="pre">spac.data_utils.</span></span><span class="sig-name descname"><span class="pre">bin2cat</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">one_hot_annotations</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">new_annotation</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/data_utils.html#bin2cat"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.data_utils.bin2cat" title="Permalink to this definition"></a></dt>
<dd><p>Combine a set of columns representing
a binary one hot encoding of categories
into a new categorical column.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>pandas.DataFrame</em>) – The pandas dataframe containing the one hot encoded annotations.</p></li>
<li><p><strong>one_hot_annotations</strong> (<em>str</em><em> or </em><em>list</em><em> of </em><em>str</em>) – A string or a list of strings representing
python regular expression of the one hot encoded annotations
columns in the data frame.</p></li>
<li><p><strong>new_annotation</strong> (<em>str</em>) – The column name for new categorical annotation to be created.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>pandas.DataFrame</em> – DataFrame with new categorical column added.</p></li>
<li><p><em>Example</em></p></li>
<li><p><em>——–</em></p></li>
<li><p><em>&gt;&gt;&gt; data = pd.DataFrame({</em></p></li>
<li><p><strong>…    ‘A’</strong> (<em>[1, 1, 0, 0],</em>)</p></li>
<li><p><strong>…     ‘B’</strong> (<em>[0, 0, 1, 0]</em>)</p></li>
<li><p><em>… })</em></p></li>
<li><p><em>&gt;&gt;&gt; one_hot_annotations = [‘A’, ‘B’]</em></p></li>
<li><p><em>&gt;&gt;&gt; new_annotation = ‘new_category’</em></p></li>
<li><p><em>&gt;&gt;&gt; result = bin2cat(data, one_hot_annotations, new_annotation)</em></p></li>
<li><p><em>&gt;&gt;&gt; print(result[new_annotation])</em></p></li>
<li><p><em>0      A</em></p></li>
<li><p><em>1      A</em></p></li>
<li><p><em>2      B</em></p></li>
<li><p><em>3    NaN</em></p></li>
<li><p><strong>Name</strong> (<em>new_category, dtype: object</em>)</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.data_utils.calculate_centroid">
<span class="sig-prename descclassname"><span class="pre">spac.data_utils.</span></span><span class="sig-name descname"><span class="pre">calculate_centroid</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x_min</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x_max</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y_min</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y_max</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">new_x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">new_y</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/data_utils.html#calculate_centroid"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.data_utils.calculate_centroid" title="Permalink to this definition"></a></dt>
<dd><p>Calculate the spatial coordinates of the cell centroid as the average of
min and max coordinates.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>pd.DataFrame</em>) – The input data frame. The dataframe should contain four columns for
x_min, x_max, y_min, and y_max for centroid calculation.</p></li>
<li><p><strong>x_min</strong> (<em>str</em>) – column name with minimum x value</p></li>
<li><p><strong>x_max</strong> (<em>str</em>) – column name with maximum x value</p></li>
<li><p><strong>y_min</strong> (<em>str</em>) – column name with minimum y value</p></li>
<li><p><strong>y_max</strong> (<em>str</em>) – column name with maximum y value</p></li>
<li><p><strong>new_x</strong> (<em>str</em>) – the new column name of the x dimension of the cientroid,
allowing characters are alphabetic, digits and underscore</p></li>
<li><p><strong>new_y</strong> (<em>str</em>) – the new column name of the y dimension of the centroid,
allowing characters are alphabetic, digits and underscore</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>data</strong> – dataframe with two new centroid columns addded. Note that the
dataframe is modified in place.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.data_utils.combine_annotations">
<span class="sig-prename descclassname"><span class="pre">spac.data_utils.</span></span><span class="sig-name descname"><span class="pre">combine_annotations</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotations</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">separator</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">new_annotation_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">AnnData</span></span></span><a class="reference internal" href="_modules/spac/data_utils.html#combine_annotations"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.data_utils.combine_annotations" title="Permalink to this definition"></a></dt>
<dd><p>Combine multiple annotations into a new annotation using a defined separator.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – The input AnnData object whose .obs will be modified.</p></li>
<li><p><strong>annotations</strong> (<em>list</em>) – List of annotation column names to combine.</p></li>
<li><p><strong>separator</strong> (<em>str</em>) – Separator to use when combining annotations.</p></li>
<li><p><strong>new_annotation_name</strong> (<em>str</em>) – The name of the new annotation to be created.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The AnnData object with the combined annotation added.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>AnnData</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.data_utils.combine_dfs">
<span class="sig-prename descclassname"><span class="pre">spac.data_utils.</span></span><span class="sig-name descname"><span class="pre">combine_dfs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dataframes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/data_utils.html#combine_dfs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.data_utils.combine_dfs" title="Permalink to this definition"></a></dt>
<dd><p>Combined multiple pandas dataframes into one.
Schema of the first dataframe is considered primary.
A warming will be printed if schema of current dataframe
is different than the primary.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>dataframes</strong> (<em>list</em><em>[</em><em>pd.DataFrame</em><em>]</em>) – A list of pandas dataframe to be combined</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>A pd.DataFrame of combined dataframs.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.data_utils.concatinate_regions">
<span class="sig-prename descclassname"><span class="pre">spac.data_utils.</span></span><span class="sig-name descname"><span class="pre">concatinate_regions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">regions</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/data_utils.html#concatinate_regions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.data_utils.concatinate_regions" title="Permalink to this definition"></a></dt>
<dd><p>Concatinate data from multiple regions and create new indexes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>regions</strong> (<em>list</em><em> of </em><em>anndata.AnnData</em>) – AnnData objects to be concatinated.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>New AnddData object with the concatinated values in AnnData.X</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>anndata.AnnData</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.data_utils.downsample_cells">
<span class="sig-prename descclassname"><span class="pre">spac.data_utils.</span></span><span class="sig-name descname"><span class="pre">downsample_cells</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotations</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_samples</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stratify</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">combined_col_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'_combined_'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/data_utils.html#downsample_cells"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.data_utils.downsample_cells" title="Permalink to this definition"></a></dt>
<dd><p>Custom downsampling of data based on one or more annotations.</p>
<p>This function offers two primary modes of operation:
1. <strong>Grouping (stratify=False)</strong>:</p>
<blockquote>
<div><ul class="simple">
<li><p>For a single annotation: The data is grouped by unique values of the
annotation, and ‘n_samples’ rows are selected from each group.</p></li>
<li><p>For multiple annotations: The data is grouped based on unique
combinations of the annotations, and ‘n_samples’ rows are selected
from each combined group.</p></li>
</ul>
</div></blockquote>
<ol class="arabic" start="2">
<li><p><strong>Stratification (stratify=True)</strong>:
- Annotations (single or multiple) are combined into a new column.
- Proportionate stratified sampling is performed based on the unique</p>
<blockquote>
<div><p>combinations in the new column, ensuring that the downsampled dataset
maintains the proportionate representation of each combined group
from the original dataset.</p>
</div></blockquote>
</li>
</ol>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>input_data</strong> (<em>pd.DataFrame</em>) – The input data frame.</p></li>
<li><p><strong>annotations</strong> (<em>str</em><em> or </em><em>list</em><em> of </em><em>str</em>) – The column name(s) to downsample on. If multiple column names are
provided, their values are combined using an underscore as a separator.</p></li>
<li><p><strong>n_samples</strong> (<em>int</em><em>, </em><em>default=None</em>) – <p>The number of samples to return. Behavior differs based on the
‘stratify’ parameter:
- stratify=False: Returns ‘n_samples’ for each unique value (or</p>
<blockquote>
<div><p>combination) of annotations.</p>
</div></blockquote>
<ul>
<li><p>stratify=True: Returns a total of ‘n_samples’ stratified by the
frequency of every label or combined labels in the annotation(s).</p></li>
</ul>
</p></li>
<li><p><strong>stratify</strong> (<em>bool</em><em>, </em><em>default=False</em>) – If true, perform proportionate stratified sampling based on the unique
combinations of annotations. This ensures that the downsampled dataset
maintains the proportionate representation of each combined group from
the original dataset.</p></li>
<li><p><strong>rand</strong> (<em>bool</em><em>, </em><em>default=False</em>) – If true and stratify is True, randomly select the returned cells.
Otherwise, choose the first n cells.</p></li>
<li><p><strong>combined_col_name</strong> (<em>str</em><em>, </em><em>default='_combined_'</em>) – Name of the column that will store combined values when multiple
annotation columns are provided.</p></li>
<li><p><strong>min_threshold</strong> (<em>int</em><em>, </em><em>default=5</em>) – The minimum number of samples a combined group should have in the
original dataset to be considered in the downsampled dataset. Groups
with fewer samples than this threshold will be excluded from the
stratification process. Adjusting this parameter determines the
minimum presence a combined group should have in the original dataset
to appear in the downsampled version.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>output_data</strong> – The proportionately stratified downsampled data frame.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>This function emphasizes proportionate stratified sampling, ensuring that
the downsampled dataset is a representative subset of the original data
with respect to the combined annotations. Due to this proportionate nature,
not all unique combinations from the original dataset might be present in
the downsampled dataset, especially if a particular combination has very
few samples in the original dataset. The <cite>min_threshold</cite> parameter can be
adjusted to determine the minimum number of samples a combined group
should have in the original dataset to appear in the downsampled version.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.data_utils.ingest_cells">
<span class="sig-prename descclassname"><span class="pre">spac.data_utils.</span></span><span class="sig-name descname"><span class="pre">ingest_cells</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dataframe</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">regex_str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/data_utils.html#ingest_cells"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.data_utils.ingest_cells" title="Permalink to this definition"></a></dt>
<dd><p>Read the csv file into an anndata object.</p>
<p>The function will also intialize features and spatial coordiantes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dataframe</strong> (<em>pandas.DataFrame</em>) – The data frame that contains cells as rows, and cells informations as
columns.</p></li>
<li><p><strong>regex_str</strong> (<em>str</em><em> or </em><em>list</em><em> of </em><em>str</em>) – A string or a list of strings representing python regular expression
for the features columns in the data frame.  x_col : str The column
name for the x coordinate of the cell.</p></li>
<li><p><strong>y_col</strong> (<em>str</em>) – The column name for the y coordinate of the cell.</p></li>
<li><p><strong>annotation</strong> (<em>str</em><em> or </em><em>list</em><em> of </em><em>str</em>) – The column name for the region that the cells. If a list is passed,
multiple annotations will be created in the returned AnnData object.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The generated AnnData object</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>anndata.AnnData</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.data_utils.load_csv_files">
<span class="sig-prename descclassname"><span class="pre">spac.data_utils.</span></span><span class="sig-name descname"><span class="pre">load_csv_files</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_names</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/data_utils.html#load_csv_files"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.data_utils.load_csv_files" title="Permalink to this definition"></a></dt>
<dd><p>Read the csv file(s) into a pandas dataframe.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>file_names</strong> (<em>str</em><em> or </em><em>list</em>) – A list of csv file paths to be
combined into single list of dataframe output</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A pandas dataframe of all the csv files. The returned dataset
will have an extra column called “loaded_file_name” containing
source file name.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.dataframe</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.data_utils.rescale_features">
<span class="sig-prename descclassname"><span class="pre">spac.data_utils.</span></span><span class="sig-name descname"><span class="pre">rescale_features</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">features</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_quantile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.01</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_quantile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.99</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/data_utils.html#rescale_features"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.data_utils.rescale_features" title="Permalink to this definition"></a></dt>
<dd><p>Clip and rescale features outside the minimum and maximum quantile.</p>
<p>The rescaled features will be between 0 and 1.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>features</strong> (<em>pandas.Dataframe</em>) – The DataRrame of features.</p></li>
<li><p><strong>min_quantile</strong> (<em>float</em>) – The minimum quantile to be consider zero.</p></li>
<li><p><strong>max_quantile</strong> (<em>float</em>) – The maximum quantile to be considerd 1.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The created DataFrame with normalized features.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.data_utils.select_values">
<span class="sig-prename descclassname"><span class="pre">spac.data_utils.</span></span><span class="sig-name descname"><span class="pre">select_values</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">values</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude_values</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/data_utils.html#select_values"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.data_utils.select_values" title="Permalink to this definition"></a></dt>
<dd><p>Selects values from either a pandas DataFrame or an AnnData object based
on the annotation and values.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>pandas.DataFrame</em><em> or </em><em>anndata.AnnData</em>) – The input data. Can be a DataFrame for tabular data or an AnnData
object.</p></li>
<li><p><strong>annotation</strong> (<em>str</em>) – The column name in a DataFrame or the annotation key in an AnnData
object to be used for selection.</p></li>
<li><p><strong>values</strong> (<em>str</em><em> or </em><em>list</em><em> of </em><em>str</em>) – List of values for the annotation to include. If None, all values are
considered for selection.</p></li>
<li><p><strong>exclude_values</strong> (<em>str</em><em> or </em><em>list</em><em> of </em><em>str</em>) – List of values for the annotation to exclude. Can’t be combined with
values.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The filtered DataFrame or AnnData object containing only the selected
rows based on the annotation and values.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame or anndata.AnnData</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.data_utils.subtract_min_per_region">
<span class="sig-prename descclassname"><span class="pre">spac.data_utils.</span></span><span class="sig-name descname"><span class="pre">subtract_min_per_region</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_quantile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.01</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/data_utils.html#subtract_min_per_region"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.data_utils.subtract_min_per_region" title="Permalink to this definition"></a></dt>
<dd><p>Substract the minimum quantile of every marker per region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>anndata.AnnData</em>) – The AnnData object.</p></li>
<li><p><strong>annotation</strong> (<em>str</em>) – The name of the annotation in <cite>adata</cite> to define batches.</p></li>
<li><p><strong>min_quantile</strong> (<em>float</em>) – The minimum quantile to rescale to zero.</p></li>
<li><p><strong>layer</strong> (<em>str</em>) – The name of the new layer to add to the AnnData object.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.data_utils.subtract_min_quantile">
<span class="sig-prename descclassname"><span class="pre">spac.data_utils.</span></span><span class="sig-name descname"><span class="pre">subtract_min_quantile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">features</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_quantile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.01</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/data_utils.html#subtract_min_quantile"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.data_utils.subtract_min_quantile" title="Permalink to this definition"></a></dt>
<dd><p>Subtract the features defined by the minimum quantile from all columns.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>features</strong> (<em>pandas.DataFrame</em>) – The dataframe of features.</p></li>
<li><p><strong>min_quantile</strong> (<em>float</em>) – The minimum quantile to be consider zero.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>dataframe with rescaled features.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.data_utils.summarize_dataframe">
<span class="sig-prename descclassname"><span class="pre">spac.data_utils.</span></span><span class="sig-name descname"><span class="pre">summarize_dataframe</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">columns</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">print_nan_locations</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">dict</span></span></span><a class="reference internal" href="_modules/spac/data_utils.html#summarize_dataframe"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.data_utils.summarize_dataframe" title="Permalink to this definition"></a></dt>
<dd><p>Summarize specified columns in a DataFrame.</p>
<p>For numeric columns, computes summary statistics.
For categorical columns, returns unique labels and frequencies.
In both cases, missing values (None/NaN) are flagged and their row indices
identified.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pd.DataFrame</em>) – The DataFrame to summarize.</p></li>
<li><p><strong>columns</strong> (<em>str</em><em> or </em><em>list</em><em> of </em><em>str</em>) – The column name or list of column names to analyze.</p></li>
<li><p><strong>print_nan_locations</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, prints the row indices where None/NaN values occur.
Default is False.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p>A dictionary where each key is a column name and its value is another
dictionary with:</p>
<blockquote>
<div><ul class="simple">
<li><p>’data_type’: either ‘numeric’ or ‘categorical’</p></li>
<li><p>’missing_count’: int</p></li>
<li><p>’missing_indices’: list of row indices with missing values</p></li>
<li><p>’summary’: summary statistics if numeric or unique labels with</p></li>
</ul>
<p>counts if categorical</p>
</div></blockquote>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<span class="target" id="module-spac.phenotyping"></span><dl class="py function">
<dt class="sig sig-object py" id="spac.phenotyping.apply_phenotypes">
<span class="sig-prename descclassname"><span class="pre">spac.phenotyping.</span></span><span class="sig-name descname"><span class="pre">apply_phenotypes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">phenotypes_dic</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/phenotyping.html#apply_phenotypes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.phenotyping.apply_phenotypes" title="Permalink to this definition"></a></dt>
<dd><p>Add binary columns to the DataFrame indicating if each cell matches a
phenotype.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_df</strong> (<em>pandas.DataFrame</em>) – The DataFrame to which binary phenotype columns will be added.</p></li>
<li><p><strong>phenotypes_dic</strong> (<em>dict</em>) – A dictionary where the keys are phenotype names and the values are
dictionaries mapping column names to values.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A dictionary where the keys are phenotype names and the values are
the counts of rows that match each phenotype.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>The function creates binary columns in the DataFrame for each phenotype
and counts the number of rows matching each phenotype.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.phenotyping.assign_manual_phenotypes">
<span class="sig-prename descclassname"><span class="pre">spac.phenotyping.</span></span><span class="sig-name descname"><span class="pre">assign_manual_phenotypes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">phenotypes_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'manual_phenotype'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">suffix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">multiple</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">drop_binary_code</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/phenotyping.html#assign_manual_phenotypes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.phenotyping.assign_manual_phenotypes" title="Permalink to this definition"></a></dt>
<dd><p>Assign manual phenotypes to the DataFrame and generate summaries.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_df</strong> (<em>pandas.DataFrame</em>) – The DataFrame to which manual phenotypes will be assigned.</p></li>
<li><p><strong>phenotypes_df</strong> (<em>pandas.DataFrame</em>) – <p>A DataFrame containing phenotype definitions with columns:
- “phenotype_name” : str</p>
<blockquote>
<div><p>The name of the phenotype.</p>
</div></blockquote>
<ul>
<li><dl class="simple">
<dt>”phenotype_code”<span class="classifier">str</span></dt><dd><p>The code used to decode the phenotype.</p>
</dd>
</dl>
</li>
</ul>
</p></li>
<li><p><strong>annotation</strong> (<em>str</em><em>, </em><em>optional</em>) – The name of the column to store the combined phenotype. Default is
“manual_phenotype”.</p></li>
<li><p><strong>prefix</strong> (<em>str</em><em>, </em><em>optional</em>) – Prefix to be added to the column names. Default is ‘’.</p></li>
<li><p><strong>suffix</strong> (<em>str</em><em>, </em><em>optional</em>) – Suffix to be added to the column names. Default is ‘’.</p></li>
<li><p><strong>multiple</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to concatenate the names of multiple positive phenotypes.
Default is True.</p></li>
<li><p><strong>drop_binary_code</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to drop the binary phenotype columns. Default is True.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p>A dictionary with the following keys:
- “phenotypes_counts”: dict</p>
<blockquote>
<div><p>Counts of cells matching each defined phenotype.</p>
</div></blockquote>
<ul class="simple">
<li><dl class="simple">
<dt>”assigned_phenotype_counts”: dict</dt><dd><p>Counts of cells matching different numbers of phenotypes.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>”multiple_phenotypes_summary”: pandas.DataFrame</dt><dd><p>Summary of cells with multiple phenotypes.</p>
</dd>
</dl>
</li>
</ul>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>The function generates a combined phenotype column, prints summaries of
cells matching multiple phenotypes, and returns a dictionary with
detailed counts and summaries.</p>
<p class="rubric">Examples</p>
<p>Suppose <cite>data_df</cite> is a DataFrame with binary phenotype columns and
<cite>phenotypes_df</cite> contains the following definitions:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">data_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span>
<span class="gp">... </span>    <span class="s1">&#39;cd4_phenotype&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span>
<span class="gp">... </span>    <span class="s1">&#39;cd8_phenotype&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span>
<span class="gp">... </span><span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">phenotypes_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">([</span>
<span class="gp">... </span>    <span class="p">{</span><span class="s2">&quot;phenotype_name&quot;</span><span class="p">:</span> <span class="s2">&quot;cd4_cells&quot;</span><span class="p">,</span> <span class="s2">&quot;phenotype_code&quot;</span><span class="p">:</span> <span class="s2">&quot;cd4+&quot;</span><span class="p">},</span>
<span class="gp">... </span>    <span class="p">{</span><span class="s2">&quot;phenotype_name&quot;</span><span class="p">:</span> <span class="s2">&quot;cd8_cells&quot;</span><span class="p">,</span> <span class="s2">&quot;phenotype_code&quot;</span><span class="p">:</span> <span class="s2">&quot;cd8+&quot;</span><span class="p">},</span>
<span class="gp">... </span>    <span class="p">{</span><span class="s2">&quot;phenotype_name&quot;</span><span class="p">:</span> <span class="s2">&quot;cd4_cd8&quot;</span><span class="p">,</span> <span class="s2">&quot;phenotype_code&quot;</span><span class="p">:</span> <span class="s2">&quot;cd4+cd8+&quot;</span><span class="p">}</span>
<span class="gp">... </span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">result</span> <span class="o">=</span> <span class="n">assign_manual_phenotypes</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">data_df</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">phenotypes_df</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">annotation</span><span class="o">=</span><span class="s2">&quot;manual&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">prefix</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">suffix</span><span class="o">=</span><span class="s1">&#39;_phenotype&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">multiple</span><span class="o">=</span><span class="kc">True</span>
<span class="gp">... </span><span class="p">)</span>
</pre></div>
</div>
<p>The <cite>data_df</cite> DataFrame will be edited in place to include a new column
<cite>“manual”</cite> with the combined phenotype labels:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">data_df</span><span class="p">)</span>
<span class="go">   cd4_phenotype  cd8_phenotype manual</span>
<span class="go">0              0              0 no_label</span>
<span class="go">1              1              0 cd4_cells</span>
<span class="go">2              0              1 cd8_cells</span>
<span class="go">3              1              1 cd8_cells, cd4_cd8</span>
</pre></div>
</div>
<p>The result dictionary contains counts and summaries as follows:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="s2">&quot;phenotypes_counts&quot;</span><span class="p">])</span>
<span class="go">{&#39;cd4_cells&#39;: 1, &#39;cd8_cells&#39;: 2, &#39;cd4_cd8&#39;: 1}</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="s2">&quot;assigned_phenotype_counts&quot;</span><span class="p">])</span>
<span class="go">0    1</span>
<span class="go">1    2</span>
<span class="go">2    1</span>
<span class="go">Name: num_phenotypes, dtype: int64</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="s2">&quot;multiple_phenotypes_summary&quot;</span><span class="p">])</span>
<span class="go">               manual  count</span>
<span class="go">0  cd8_cells, cd4_cd8      1</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.phenotyping.combine_phenotypes">
<span class="sig-prename descclassname"><span class="pre">spac.phenotyping.</span></span><span class="sig-name descname"><span class="pre">combine_phenotypes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">phenotype_columns</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">multiple</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/phenotyping.html#combine_phenotypes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.phenotyping.combine_phenotypes" title="Permalink to this definition"></a></dt>
<dd><p>Combine multiple binary phenotype columns into a new column in a vectorized manner.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_df</strong> (<em>pandas.DataFrame</em>) – DataFrame containing the phenotype columns.</p></li>
<li><p><strong>phenotype_columns</strong> (<em>list</em><em> of </em><em>str</em>) – List of binary phenotype column names.</p></li>
<li><p><strong>multiple</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to concatenate the names of multiple positive phenotypes.
If False, all multiple positive phenotypes are labeled as
“no_label”. Default is True.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A Series representing the combined phenotype for each row.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.Series</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.phenotyping.decode_phenotype">
<span class="sig-prename descclassname"><span class="pre">spac.phenotyping.</span></span><span class="sig-name descname"><span class="pre">decode_phenotype</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">phenotype_code</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/phenotyping.html#decode_phenotype"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.phenotyping.decode_phenotype" title="Permalink to this definition"></a></dt>
<dd><p>Convert a phenotype code into a dictionary mapping
feature (marker) names to values for that marker’s
classification as ‘+’ or ‘-‘.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>pandas.DataFrame</em>) – The DataFrame containing the columns that will be used to decode
the phenotype.</p></li>
<li><p><strong>phenotype_code</strong> (<em>str</em>) – The phenotype code string, which should end with ‘+’ or ‘-‘.</p></li>
<li><p><strong>**kwargs</strong> (<em>keyword arguments</em>) – <p>Optional keyword arguments to specify prefix and suffix to be added
to the column names.
- prefix : str, optional</p>
<blockquote>
<div><p>Prefix to be added to the column names for the
feature classification. Default is ‘’.</p>
</div></blockquote>
<ul>
<li><dl class="simple">
<dt>suffix<span class="classifier">str, optional</span></dt><dd><p>Suffix to be added to the column names for the
feature classification. Default is ‘’.</p>
</dd>
</dl>
</li>
</ul>
</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A dictionary where the keys are column names and the values are the
corresponding phenotype classification.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If the phenotype code does not end with ‘+’ or ‘-’ or if any columns
    specified in the phenotype code do not exist in the DataFrame.</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>The function splits the phenotype code on ‘+’ and ‘-’ characters to
determine the phenotype columns and values. It checks if the columns
exist in the DataFrame and whether they are binary or string types to
properly map values.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.phenotyping.generate_phenotypes_dict">
<span class="sig-prename descclassname"><span class="pre">spac.phenotyping.</span></span><span class="sig-name descname"><span class="pre">generate_phenotypes_dict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">phenotypes_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">suffix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/phenotyping.html#generate_phenotypes_dict"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.phenotyping.generate_phenotypes_dict" title="Permalink to this definition"></a></dt>
<dd><p>Generate a dictionary of phenotype names to their corresponding
decoding rules.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_df</strong> (<em>pandas.DataFrame</em>) – The DataFrame containing the columns that will be used to decode
the phenotypes.</p></li>
<li><p><strong>phenotypes_df</strong> (<em>pandas.DataFrame</em>) – <p>A DataFrame containing phenotype definitions with columns:
- “phenotype_name” : str</p>
<blockquote>
<div><p>The name of the phenotype.</p>
</div></blockquote>
<ul>
<li><dl class="simple">
<dt>”phenotype_code”<span class="classifier">str</span></dt><dd><p>The code used to decode the phenotype.</p>
</dd>
</dl>
</li>
</ul>
</p></li>
<li><p><strong>prefix</strong> (<em>str</em><em>, </em><em>optional</em>) – Prefix to be added to the column names. Default is ‘’.</p></li>
<li><p><strong>suffix</strong> (<em>str</em><em>, </em><em>optional</em>) – Suffix to be added to the column names. Default is ‘’.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A dictionary where the keys are phenotype names and the values are
dictionaries mapping column names to values.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>The function iterates over each row in the <cite>phenotypes_df</cite> DataFrame and
decodes the phenotype using the <cite>decode_phenotype</cite> function.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.phenotyping.is_binary_0_1">
<span class="sig-prename descclassname"><span class="pre">spac.phenotyping.</span></span><span class="sig-name descname"><span class="pre">is_binary_0_1</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">column</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/phenotyping.html#is_binary_0_1"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.phenotyping.is_binary_0_1" title="Permalink to this definition"></a></dt>
<dd><p>Check if a pandas Series contains only binary values (0 and 1).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>column</strong> (<em>pandas.Series</em>) – The pandas Series to check.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>True if the Series contains only 0 and 1, False otherwise.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>The function considers a Series to be binary if it contains exactly
the values 0 and 1, and no other values.</p>
</dd></dl>

<span class="target" id="module-spac.spatial_analysis"></span><dl class="py function">
<dt class="sig sig-object py" id="spac.spatial_analysis.calculate_nearest_neighbor">
<span class="sig-prename descclassname"><span class="pre">spac.spatial_analysis.</span></span><span class="sig-name descname"><span class="pre">calculate_nearest_neighbor</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spatial_associated_table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'spatial'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">imageid</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">label</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'spatial_distance'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/spatial_analysis.html#calculate_nearest_neighbor"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.spatial_analysis.calculate_nearest_neighbor" title="Permalink to this definition"></a></dt>
<dd><p>Computes the shortest distance from each cell to the nearest cell of
each phenotype (via scimap.tl.spatial_distance) and stores the
resulting DataFrame in <cite>adata.obsm[label]</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>anndata.AnnData</em>) – Annotated data matrix with spatial information.</p></li>
<li><p><strong>annotation</strong> (<em>str</em>) – Column name in <cite>adata.obs</cite> containing cell annotationsi (i.e.
phenotypes).</p></li>
<li><p><strong>spatial_associated_table</strong> (<em>str</em><em>, </em><em>optional</em>) – Key in <cite>adata.obsm</cite> where spatial coordinates are stored. Default is
‘spatial’.</p></li>
<li><p><strong>imageid</strong> (<em>str</em><em>, </em><em>optional</em>) – The column in <cite>adata.obs</cite> specifying image IDs. If None,
a dummy image column is created temporarily. Spatial distances are
computed across the entire dataseti as if it’s one image.</p></li>
<li><p><strong>label</strong> (<em>str</em><em>, </em><em>optional</em>) – The key under which results are stored  in <cite>adata.obsm</cite>. Default is
‘spatial_distance’.</p></li>
<li><p><strong>verbose</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, prints progress messages. Default is True.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Modifies <cite>adata</cite> in place by storing a DataFrame of
spatial distances in <cite>adata.obsm[label]</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
<p class="rubric">Example</p>
<p>For a dataset with two cells (CellA, CellB) both of the same phenotype
“type1”, the output might look like:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;spatial_distance&#39;</span><span class="p">]</span>
<span class="go">       type1</span>
<span class="go">CellA    0.0</span>
<span class="go">CellB    0.0</span>
</pre></div>
</div>
<p>For a dataset with two phenotypes “type1” and “type2”, the output might
look like:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;spatial_distance&#39;</span><span class="p">]</span>
<span class="go">       type1     type2</span>
<span class="go">CellA   0.00  1.414214</span>
<span class="go">CellB  1.414214  0.00</span>
</pre></div>
</div>
<dl>
<dt>Input:</dt><dd><dl class="simple">
<dt>adata.obs:</dt><dd><p>cell_type   imageid
type1       image1
type1       image1
type2       image1</p>
</dd>
<dt>adata.obsm[‘spatial’]:</dt><dd><p>[[0.0, 0.0], [1.0, 1.0], [2.0, 2.0]]</p>
</dd>
</dl>
</dd>
<dt>Output stored in adata.obsm[‘spatial_distance’]:</dt><dd><blockquote>
<div><p>type1   type2</p>
</div></blockquote>
<p>0    0.0    1.414
1    1.414  0.0
2    2.236  1.0</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ValueError</strong> – If <cite>spatial_associated_table</cite> is not found in <cite>adata.obsm</cite>.
    If spatial coordinates are missing or invalid.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.spatial_analysis.neighborhood_profile">
<span class="sig-prename descclassname"><span class="pre">spac.spatial_analysis.</span></span><span class="sig-name descname"><span class="pre">neighborhood_profile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">phenotypes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">distances</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">regions</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spatial_key</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'spatial'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">normalize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">associated_table_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'neighborhood_profile'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/spatial_analysis.html#neighborhood_profile"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.spatial_analysis.neighborhood_profile" title="Permalink to this definition"></a></dt>
<dd><p>Calculate the neighborhood profile for every cell in all slides in an analysis
and update the input AnnData object in place.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – The AnnData object containing the spatial coordinates and phenotypes.</p></li>
<li><p><strong>phenotypes</strong> (<em>str</em>) – The name of the column in adata.obs that contains the phenotypes.</p></li>
<li><p><strong>distances</strong> (<em>list</em>) – The list of increasing distances for the neighborhood profile.</p></li>
<li><p><strong>spatial_key</strong> (<em>str</em><em>, </em><em>optional</em>) – The key in adata.obs that contains the spatial coordinates. Default is
‘spatial’.</p></li>
<li><p><strong>normalize</strong> (<em>str</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – If ‘total_cells’, normalize the neighborhood profile based on the
total number of cells in each bin.
If ‘bin_area’, normalize the neighborhood profile based on the area
of every bin.  Default is None.</p></li>
<li><p><strong>associated_table_name</strong> (<em>str</em><em>, </em><em>optional</em>) – The name of the column in adata.obsm that will contain the
neighborhood profile. Default is ‘neighborhood_profile’.</p></li>
<li><p><strong>regions</strong> (<em>str</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – The name of the column in adata.obs that contains the regions.
If None, all cells in adata will be used. Default is None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The function modifies the input AnnData object in place, adding a new
column containing the neighborhood profile to adata.obsm.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>The input AnnData object ‘adata’ is modified in place. The function adds a
new column containing the neighborhood profile to adata.obsm, named by the
parameter ‘associated_table_name’. The associated_table_name is a 3D array of
shape (n_cells, n_phenotypes, n_bins) where n_cells is the number of cells
in the all slides, n_phenotypes is the number of unique phenotypes, and
n_bins is the number of bins in the distances list.</p>
<p>A dictionary is added to adata.uns[associated_table_name] with the two keys
“bins” and “labels”. “labels” will store all the values in the phenotype
annotation.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.spatial_analysis.ripley_l">
<span class="sig-prename descclassname"><span class="pre">spac.spatial_analysis.</span></span><span class="sig-name descname"><span class="pre">ripley_l</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">phenotypes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">distances</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">regions</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spatial_key</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'spatial'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_simulations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">area</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">42</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edge_correction</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/spatial_analysis.html#ripley_l"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.spatial_analysis.ripley_l" title="Permalink to this definition"></a></dt>
<dd><p>Calculate Ripley’s L statistic for spatial data in <cite>adata</cite>.</p>
<p>Ripley’s L statistic is a spatial point pattern analysis metric that
quantifies clustering or regularity in point patterns across various
distances. This function calculates the statistic for each region in
<cite>adata</cite> (if provided) or for all cells if regions are not specified.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – The annotated data matrix containing the spatial coordinates and
cell phenotypes.</p></li>
<li><p><strong>annotation</strong> (<em>str</em>) – The key in <cite>adata.obs</cite> representing the annotation for cell phenotypes.</p></li>
<li><p><strong>phenotypes</strong> (<em>list</em><em> of </em><em>str</em>) – A list containing two phenotypes for which the Ripley L statistic
will be calculated. If the two phenotypes are the same, the
calculation is done for the same type; if different, it considers
interactions between the two.</p></li>
<li><p><strong>distances</strong> (<em>array-like</em>) – An array of distances at which to calculate Ripley’s L statistic.
The values must be positive and incremental.</p></li>
<li><p><strong>regions</strong> (<em>str</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – The key in <cite>adata.obs</cite> representing regions for stratifying the
data. If <cite>None</cite>, all cells will be treated as one region.</p></li>
<li><p><strong>spatial_key</strong> (<em>str</em><em>, </em><em>optional</em>) – The key in <cite>adata.obsm</cite> representing the spatial coordinates.
Default is <cite>“spatial”</cite>.</p></li>
<li><p><strong>n_simulations</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of simulations to perform for significance testing. Default is 100.</p></li>
<li><p><strong>area</strong> (<em>float</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – The area of the spatial region of interest. If <cite>None</cite>, the area
will be inferred from the data. Default is <cite>None</cite>.</p></li>
<li><p><strong>seed</strong> (<em>int</em><em>, </em><em>optional</em>) – Random seed for simulation reproducibility. Default is 42.</p></li>
<li><p><strong>edge_correction</strong> (<em>boo</em><em>, </em><em>optional</em>) – If True, apply edge correction to the Ripley’s L calculation.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A DataFrame containing the Ripley’s L results for each region or
the entire dataset if <cite>regions</cite> is <cite>None</cite>. The DataFrame includes
the following columns:
- <cite>region</cite>: The region label or ‘all’ if no regions are specified.
- <cite>center_phenotype</cite>: The first phenotype in <cite>phenotypes</cite>.
- <cite>neighbor_phenotype</cite>: The second phenotype in <cite>phenotypes</cite>.
- <cite>ripley_l</cite>: The Ripley’s L statistic calculated for the region.
- <cite>config</cite>: A dictionary with configuration settings used for the calculation.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>Ripley’s L is an adjusted version of Ripley’s K that corrects for the
inherent increase in point-to-point distances as the distance grows.
This statistic is used to evaluate spatial clustering or dispersion
of points (cells) in biological datasets.</p>
<p>The function uses pre-defined distances and performs simulations to
assess the significance of observed patterns. The results are stored
in the <cite>.uns</cite> attribute of <cite>adata</cite> under the key <cite>‘ripley_l’</cite>, or in
a new DataFrame if no prior results exist.</p>
<p class="rubric">Examples</p>
<p>Calculate Ripley’s L for two phenotypes in a single region dataset:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">result</span> <span class="o">=</span> <span class="n">ripley_l</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">annotation</span><span class="o">=</span><span class="s1">&#39;cell_type&#39;</span><span class="p">,</span> <span class="n">phenotypes</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">],</span> <span class="n">distances</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">500</span><span class="p">,</span> <span class="mi">100</span><span class="p">))</span>
</pre></div>
</div>
<p>Calculate Ripley’s L for multiple regions in <cite>adata</cite>:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">result</span> <span class="o">=</span> <span class="n">ripley_l</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">annotation</span><span class="o">=</span><span class="s1">&#39;cell_type&#39;</span><span class="p">,</span> <span class="n">phenotypes</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">],</span> <span class="n">distances</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">500</span><span class="p">,</span> <span class="mi">100</span><span class="p">),</span> <span class="n">regions</span><span class="o">=</span><span class="s1">&#39;region_key&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.spatial_analysis.spatial_interaction">
<span class="sig-prename descclassname"><span class="pre">spac.spatial_analysis.</span></span><span class="sig-name descname"><span class="pre">spatial_interaction</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">analysis_method</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stratify_by</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_matrix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">coord_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_rings</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_neighs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">6</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">radius</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'seismic'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/spatial_analysis.html#spatial_interaction"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.spatial_analysis.spatial_interaction" title="Permalink to this definition"></a></dt>
<dd><p>Perform spatial analysis on the selected annotation in the dataset.
Current analysis methods are provided in squidpy:</p>
<blockquote>
<div><p>Neighborhood Enrichment,
Cluster Interaction Matrix</p>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>anndata.AnnData</em>) – The AnnData object.</p></li>
<li><p><strong>annotation</strong> (<em>str</em>) – The column name of the annotation (e.g., phenotypes) to analyze in the
provided dataset.</p></li>
<li><p><strong>analysis_method</strong> (<em>str</em>) – The analysis method to use, currently available:
“Neighborhood Enrichment” and “Cluster Interaction Matrix”.</p></li>
<li><p><strong>stratify_by</strong> (<em>str</em><em> or </em><em>list</em><em> of </em><em>strs</em>) – The annotation[s] to stratify the dataset when generating
interaction plots. If single annotation is passed, the dataset
will be stratified by the unique labels in the annotation column.
If n (n&gt;=2) annotations are passed, the function will be stratified
based on existing combination of labels in the passed annotations.</p></li>
<li><p><strong>ax</strong> (<em>matplotlib.axes.Axes</em><em>, </em><em>default None</em>) – The matplotlib Axes to display the image. This option is only
available when stratify is None.</p></li>
<li><p><strong>return_matrix</strong> (<em>boolean</em><em>, </em><em>default False</em>) – If true, the fucntion will return a list of two dictionaries,
the first contains axes and the second containing computed matrix.
Note that for Neighborhood Encrichment, the matrix will be a tuple
with the z-score and the enrichment count.
For Cluster Interaction Matrix, it will returns the
interaction matrix.
If False, the function will return only the axes dictionary.</p></li>
<li><p><strong>seed</strong> (<em>int</em><em>, </em><em>default None</em>) – Random seed for reproducibility, used in Neighborhood Enrichment
Analysis.</p></li>
<li><p><strong>coord_type</strong> (<em>str</em><em>, </em><em>optional</em>) – Type of coordinate system used in sq.gr.spatial_neighbors.
Should be either ‘grid’ (Visium Data) or ‘generic’ (Others).
Default is None, decided by the squidy pacakge. If spatial_key
is in anndata.uns the coord_type would be ‘grid’, otherwise
general.</p></li>
<li><p><strong>n_rings</strong> (<em>int</em><em>, </em><em>default 1</em>) – Number of rings of neighbors for grid data.
Only used when coord_type = ‘grid’ (Visium)</p></li>
<li><p><strong>n_neights</strong> (<em>int</em><em>, </em><em>optional</em>) – Default is 6.
Depending on the <code class="docutils literal notranslate"><span class="pre">coord_type</span></code>:
- ‘grid’ (Visium) - number of neighboring tiles.
- ‘generic’ - number of neighborhoods for non-grid data.</p></li>
<li><p><strong>radius</strong> (<em>float</em><em>, </em><em>optional</em>) – <p>Default is None.
Only available when coord_type = ‘generic’. Depending on the type:
- <code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code> - compute the graph based on neighborhood radius.
- <code class="xref py py-class docutils literal notranslate"><span class="pre">tuple</span></code> - prune the final graph to only contain</p>
<blockquote>
<div><p>edges in interval <cite>[min(radius), max(radius)]</cite>.</p>
</div></blockquote>
</p></li>
<li><p><strong>cmap</strong> (<em>str</em><em>, </em><em>default 'seismic'</em>) – The colormap to use for the plot. The ‘seismic’ color map
consist of three color regions: red for positive, blue for negative,
and the white at the center. This color map effectively represents
the nature of the spatial interaction analysis results,
where positive values indicate clustering and
negative values indicate seperation. For more color maps, please visit
<a class="reference external" href="https://matplotlib.org/stable/tutorials/colors/colormaps.html">https://matplotlib.org/stable/tutorials/colors/colormaps.html</a></p></li>
<li><p><strong>**kwargs</strong> – Keyword arguments for matplotlib.pyplot.text()</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p>A dictionary containing the results of the spatial interaction
analysis. The keys of the dictionary depend on the parameters
passed to the function:</p>
<dl class="simple">
<dt>Ax<span class="classifier">dict or matplotlib.axes.Axes</span></dt><dd><p>If <cite>stratify_by</cite> is not used, returns a single
matplotlib.axes.Axes object. If <cite>stratify_by</cite> is used,
returns a dictionary of Axes objects, with keys
representing the stratification groups.</p>
</dd>
<dt>Matrix<span class="classifier">dict, optional</span></dt><dd><p>Contains processed DataFrames of computed matrices with row and
column labels applied. If <cite>stratify_by</cite> is used, the keys represent
the stratification groups. For example:
- <cite>results[‘Matrix’][‘GroupA’]</cite> for a specific stratification group.
- If <cite>stratify_by</cite> is not used, the table is accessible via
<cite>results[‘Matrix’][‘annotation’]</cite>.</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<span class="target" id="module-spac.transformations"></span><dl class="py function">
<dt class="sig sig-object py" id="spac.transformations.apply_per_batch">
<span class="sig-prename descclassname"><span class="pre">spac.transformations.</span></span><span class="sig-name descname"><span class="pre">apply_per_batch</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/transformations.html#apply_per_batch"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.transformations.apply_per_batch" title="Permalink to this definition"></a></dt>
<dd><p>Apply a given function to data per batch, with additional parameters.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>np.ndarray</em>) – The data to transform.</p></li>
<li><p><strong>annotation</strong> (<em>np.ndarray</em>) – Batch annotations for each row in the data.</p></li>
<li><p><strong>method</strong> (<em>str</em>) – The function to apply to each batch. Options: ‘arcsinh_transformation’
or ‘normalize_features’.</p></li>
<li><p><strong>kwargs</strong> – Additional parameters to pass to the function.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The transformed data.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.transformations.arcsinh_transformation">
<span class="sig-prename descclassname"><span class="pre">spac.transformations.</span></span><span class="sig-name descname"><span class="pre">arcsinh_transformation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input_layer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">co_factor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">percentile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_layer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'arcsinh'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">per_batch</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/transformations.html#arcsinh_transformation"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.transformations.arcsinh_transformation" title="Permalink to this definition"></a></dt>
<dd><p>Apply arcsinh transformation using a co-factor (fixed number) or a given
percentile of each feature. This transformation can be applied to the
entire dataset or per batch based on provided parameters.</p>
<p>Computes the co-factor or percentile for each biomarker individually,
ensuring proper scaling based on its unique range of expression levels.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>anndata.AnnData</em>) – The AnnData object containing the data to transform.</p></li>
<li><p><strong>input_layer</strong> (<em>str</em><em>, </em><em>optional</em>) – The name of the layer in the AnnData object to transform.
If None, the main data matrix .X is used.</p></li>
<li><p><strong>co_factor</strong> (<em>float</em><em>, </em><em>optional</em>) – A fixed positive number to use as a co-factor for the transformation.</p></li>
<li><p><strong>percentile</strong> (<em>float</em><em>, </em><em>optional</em>) – The percentile is computed for each feature (column) individually.</p></li>
<li><p><strong>output_layer</strong> (<em>str</em><em>, </em><em>default=&quot;arcsinh&quot;</em>) – Name of the layer to put the transformed results. If it already exists,
it will be overwritten with a warning.</p></li>
<li><p><strong>per_batch</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default=False</em>) – Whether to apply the transformation per batch.</p></li>
<li><p><strong>annotation</strong> (<em>str</em><em>, </em><em>optional</em>) – The name of the annotation in <cite>adata</cite> to define batches. Required if
per_batch is True.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>adata</strong> – The AnnData object with the transformed data stored in the specified
output_layer.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>anndata.AnnData</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.transformations.arcsinh_transformation_core">
<span class="sig-prename descclassname"><span class="pre">spac.transformations.</span></span><span class="sig-name descname"><span class="pre">arcsinh_transformation_core</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">co_factor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">percentile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/transformations.html#arcsinh_transformation_core"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.transformations.arcsinh_transformation_core" title="Permalink to this definition"></a></dt>
<dd><p>Apply arcsinh transformation using a co-factori or a percentile.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>np.ndarray</em>) – The data to transform.</p></li>
<li><p><strong>co_factor</strong> (<em>float</em><em>, </em><em>optional</em>) – A fixed positive number to use as a co-factor for the
transformation.</p></li>
<li><p><strong>percentile</strong> (<em>float</em><em>, </em><em>optional</em>) – The percentile to determine the co-factor if co_factor is not
provided.
The percentile is computed for each feature (column)
individually.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The transformed data.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.ndarray</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If both co_factor and percentile are None.
    If both co_factor and percentile are specified.
    If percentile is not in the range [0, 100].</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.transformations.batch_normalize">
<span class="sig-prename descclassname"><span class="pre">spac.transformations.</span></span><span class="sig-name descname"><span class="pre">batch_normalize</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_layer</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input_layer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'median'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/transformations.html#batch_normalize"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.transformations.batch_normalize" title="Permalink to this definition"></a></dt>
<dd><p>Adjust the features of every marker using a normalization method.</p>
<p>The normalization methods are summarized here:
<a class="reference external" href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8723144/">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8723144/</a></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>anndata.AnnData</em>) – The AnnData object.</p></li>
<li><p><strong>annotation</strong> (<em>str</em>) – The name of the annotation in <cite>adata</cite> to define batches.</p></li>
<li><p><strong>output_layer</strong> (<em>str</em>) – The name of the new layer to add to the anndata object for storing
normalized data.</p></li>
<li><p><strong>input_layer</strong> (<em>str</em><em>, </em><em>optional</em>) – The name of the layer from which to read data. If None, read from <cite>.X</cite>.</p></li>
<li><p><strong>method</strong> (<em>{&quot;median&quot;</em><em>, </em><em>&quot;Q50&quot;</em><em>, </em><em>&quot;Q75&quot;</em><em>, </em><em>&quot;z-score&quot;}</em><em>, </em><em>default &quot;median&quot;</em>) – The normalization method to use.</p></li>
<li><p><strong>log</strong> (<em>bool</em><em>, </em><em>default False</em>) – If True, take the log2 of features before normalization.
Ensure this is boolean.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.transformations.get_cluster_info">
<span class="sig-prename descclassname"><span class="pre">spac.transformations.</span></span><span class="sig-name descname"><span class="pre">get_cluster_info</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">features</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/transformations.html#get_cluster_info"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.transformations.get_cluster_info" title="Permalink to this definition"></a></dt>
<dd><p>Retrieve information about clusters based on specific annotation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>anndata.AnnData</em>) – The AnnData object.</p></li>
<li><p><strong>annotation</strong> (<em>str</em>) – Annotation in adata.obs for cluster info.</p></li>
<li><p><strong>features</strong> (<em>list</em><em> of </em><em>str</em><em>, </em><em>optional</em>) – Features (e.g., markers) for cluster metrics.
Defaults to all features in adata.var_names.</p></li>
<li><p><strong>layer</strong> (<em>str</em><em>, </em><em>optional</em>) – The layer to be used in the aggregate summaries.
If None, uses adata.X.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame with metrics for each cluster including the percentage of
each cluster to the whole sample.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.transformations.knn_clustering">
<span class="sig-prename descclassname"><span class="pre">spac.transformations.</span></span><span class="sig-name descname"><span class="pre">knn_clustering</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">features</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">k</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_annotation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'knn'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">associated_table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">missing_label</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'no_label'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/transformations.html#knn_clustering"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.transformations.knn_clustering" title="Permalink to this definition"></a></dt>
<dd><p>Calculate knn clusters using sklearn KNeighborsClassifier</p>
<p>The function will add these two attributes to <cite>adata</cite>:
<cite>.obs[output_annotation]</cite></p>
<blockquote>
<div><p>The assigned int64 class labels by KNeighborsClassifier</p>
</div></blockquote>
<dl class="simple">
<dt><cite>.uns[output_annotation_features]</cite></dt><dd><p>The features used to calculate the knn clusters</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>anndata.AnnData</em>) – The AnnData object.</p></li>
<li><p><strong>features</strong> (<em>list</em><em> of </em><em>str</em>) – The variables that would be included in fitting the KNN classifier.</p></li>
<li><p><strong>annotation</strong> (<em>str</em>) – The name of the annotation used for classifying the data</p></li>
<li><p><strong>layer</strong> (<em>str</em><em>, </em><em>optional</em>) – The layer to be used.</p></li>
<li><p><strong>k</strong> (<em>int</em><em>, </em><em>optional</em>) – The number of nearest neighbor to be used in creating the graph.</p></li>
<li><p><strong>output_annotation</strong> (<em>str</em><em>, </em><em>optional</em>) – The name of the output layer where the clusters are stored.</p></li>
<li><p><strong>associated_table</strong> (<em>str</em><em>, </em><em>optional</em>) – If set, use the corresponding key <cite>adata.obsm</cite> to calcuate the
clustering. Takes priority over the layer argument.</p></li>
<li><p><strong>missing_label</strong> (<em>str</em><em> or </em><em>int</em>) – The value of missing annotations in adata.obs[annotation]</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>adata is updated inplace</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.transformations.normalize_features">
<span class="sig-prename descclassname"><span class="pre">spac.transformations.</span></span><span class="sig-name descname"><span class="pre">normalize_features</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">low_quantile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.02</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">high_quantile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.98</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interpolation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'linear'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input_layer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_layer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'normalized_feature'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">per_batch</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/transformations.html#normalize_features"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.transformations.normalize_features" title="Permalink to this definition"></a></dt>
<dd><p>Normalize the features stored in an AnnData object.
Any entry lower than the value corresponding to low_quantile of the
column will be assigned a value of low_quantile, and entry that
are greater than high_quantile value will be assigned as the value of
high_quantile.
Other entries will be normalized with
(values - quantile min)/(quantile max - quantile min).
Resulting column will have value ranged between [0, 1].</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.transformations.normalize_features_core">
<span class="sig-prename descclassname"><span class="pre">spac.transformations.</span></span><span class="sig-name descname"><span class="pre">normalize_features_core</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">low_quantile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.02</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">high_quantile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.98</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interpolation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'linear'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/transformations.html#normalize_features_core"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.transformations.normalize_features_core" title="Permalink to this definition"></a></dt>
<dd><p>Normalize the features in a numpy array.</p>
<p>Any entry lower than the value corresponding to low_quantile of the column
will be assigned a value of low_quantile, and entries that are greater than
high_quantile value will be assigned as value of high_quantile.
Other entries will be normalized with
(values - quantile min)/(quantile max - quantile min).
Resulting column will have values ranged between [0, 1].</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>np.ndarray</em>) – The data to be normalized.</p></li>
<li><p><strong>low_quantile</strong> (<em>float</em><em>, </em><em>optional</em><em> (</em><em>default: 0.02</em><em>)</em>) – The lower quantile to use for normalization. Determines the
minimum value after normalization.
Must be a positive float between [0,1).</p></li>
<li><p><strong>high_quantile</strong> (<em>float</em><em>, </em><em>optional</em><em> (</em><em>default: 0.98</em><em>)</em>) – The higher quantile to use for normalization. Determines the
maximum value after normalization.
Must be a positive float between (0,1].</p></li>
<li><p><strong>interpolation</strong> (<em>str</em><em>, </em><em>optional</em><em> (</em><em>default: &quot;linear&quot;</em><em>)</em>) – The interpolation method to use when selecting the value for
low and high quantile. Values can be “nearest” or “linear”.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The normalized data.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.ndarray</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>TypeError</strong> – If low_quantile or high_quantile are not numeric.</p></li>
<li><p><strong>ValueError</strong> – If low_quantile is not less than high_quantile, or if they are
    out of the range [0, 1] and (0, 1], respectively.</p></li>
<li><p><strong>ValueError</strong> – If interpolation is not one of the allowed values.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.transformations.phenograph_clustering">
<span class="sig-prename descclassname"><span class="pre">spac.transformations.</span></span><span class="sig-name descname"><span class="pre">phenograph_clustering</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">features</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">k</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_annotation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'phenograph'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">associated_table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/transformations.html#phenograph_clustering"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.transformations.phenograph_clustering" title="Permalink to this definition"></a></dt>
<dd><p>Calculate automatic phenotypes using phenograph.</p>
<p>The function will add these two attributes to <cite>adata</cite>:
<cite>.obs[“phenograph”]</cite></p>
<blockquote>
<div><p>The assigned int64 class by phenograph</p>
</div></blockquote>
<dl class="simple">
<dt><cite>.uns[“phenograph_features”]</cite></dt><dd><p>The features used to calculate the phenograph clusters</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>anndata.AnnData</em>) – The AnnData object.</p></li>
<li><p><strong>features</strong> (<em>list</em><em> of </em><em>str</em>) – The variables that would be included in creating the phenograph
clusters.</p></li>
<li><p><strong>layer</strong> (<em>str</em><em>, </em><em>optional</em>) – The layer to be used in calculating the phengraph clusters.</p></li>
<li><p><strong>k</strong> (<em>int</em><em>, </em><em>optional</em>) – The number of nearest neighbor to be used in creating the graph.</p></li>
<li><p><strong>seed</strong> (<em>int</em><em>, </em><em>optional</em>) – Random seed for reproducibility.</p></li>
<li><p><strong>output_annotation</strong> (<em>str</em><em>, </em><em>optional</em>) – The name of the output layer where the clusters are stored.</p></li>
<li><p><strong>associated_table</strong> (<em>str</em><em>, </em><em>optional</em>) – If set, use the corresponding key <cite>adata.obsm</cite> to calcuate the
Phenograph. Takes priority over the layer argument.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>adata</strong> – Updated AnnData object with the phenograph clusters
stored in <cite>adata.obs[output_annotation]</cite></p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>anndata.AnnData</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.transformations.rename_annotations">
<span class="sig-prename descclassname"><span class="pre">spac.transformations.</span></span><span class="sig-name descname"><span class="pre">rename_annotations</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">src_annotation</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dest_annotation</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mappings</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/transformations.html#rename_annotations"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.transformations.rename_annotations" title="Permalink to this definition"></a></dt>
<dd><p>Rename labels in a given annotation in an AnnData object based on a
provided dictionary. This function modifies the adata object in-place
and creates a new annotation column.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>anndata.AnnData</em>) – The AnnData object.</p></li>
<li><p><strong>src_annotation</strong> (<em>str</em>) – Name of the column in adata.obs containing the original
labels of the source annotation.</p></li>
<li><p><strong>dest_annotation</strong> (<em>str</em>) – The name of the new column to be created in the AnnData object
containing the renamed labels.</p></li>
<li><p><strong>mappings</strong> (<em>dict</em>) – A dictionary mapping the original annotation labels to
the new labels.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">adata</span> <span class="o">=</span> <span class="n">your_anndata_object</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">src_annotation</span> <span class="o">=</span> <span class="s2">&quot;phenograph&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mappings</span> <span class="o">=</span> <span class="p">{</span>
<span class="gp">... </span>    <span class="s2">&quot;0&quot;</span><span class="p">:</span> <span class="s2">&quot;group_8&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="s2">&quot;1&quot;</span><span class="p">:</span> <span class="s2">&quot;group_2&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="s2">&quot;2&quot;</span><span class="p">:</span> <span class="s2">&quot;group_6&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="c1"># ...</span>
<span class="gp">... </span>    <span class="s2">&quot;37&quot;</span><span class="p">:</span> <span class="s2">&quot;group_5&quot;</span><span class="p">,</span>
<span class="gp">... </span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dest_annotation</span> <span class="o">=</span> <span class="s2">&quot;renamed_annotations&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adata</span> <span class="o">=</span> <span class="n">rename_annotations</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">adata</span><span class="p">,</span> <span class="n">src_annotation</span><span class="p">,</span> <span class="n">dest_annotation</span><span class="p">,</span> <span class="n">mappings</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.transformations.run_umap">
<span class="sig-prename descclassname"><span class="pre">spac.transformations.</span></span><span class="sig-name descname"><span class="pre">run_umap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_neighbors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">75</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_dist</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_components</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metric</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'euclidean'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">random_state</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">transform_seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">42</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_derived_feature</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'X_umap'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">associated_table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/transformations.html#run_umap"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.transformations.run_umap" title="Permalink to this definition"></a></dt>
<dd><p>Perform UMAP analysis on the specific layer of the AnnData object
or the default data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – Annotated data matrix.</p></li>
<li><p><strong>n_neighbors</strong> (<em>int</em><em>, </em><em>default=75</em>) – Number of neighbors to consider when constructing the UMAP. This
influences the balance between preserving local and global structures
in the data.</p></li>
<li><p><strong>min_dist</strong> (<em>float</em><em>, </em><em>default=0.1</em>) – Minimum distance between points in the UMAP space. Controls how
tightly the embedding is allowed to compress points together.</p></li>
<li><p><strong>n_components</strong> (<em>int</em><em>, </em><em>default=2</em>) – Number of dimensions for embedding.</p></li>
<li><p><strong>metric</strong> (<em>str</em><em>, </em><em>optional</em>) – Metric to compute distances in high dimensional space.
Check <cite>https://umap-learn.readthedocs.io/en/latest/api.html</cite> for
options. The default is ‘euclidean’.</p></li>
<li><p><strong>random_state</strong> (<em>int</em><em>, </em><em>default=0</em>) – Seed used by the random number generator(RNG) during UMAP fitting.</p></li>
<li><p><strong>transform_seed</strong> (<em>int</em><em>, </em><em>default=42</em>) – RNG seed during UMAP transformation.</p></li>
<li><p><strong>layer</strong> (<em>str</em><em>, </em><em>optional</em>) – Layer of AnnData object for UMAP. Defaults to <cite>adata.X</cite>.</p></li>
<li><p><strong>output_derived_feature</strong> (<em>str</em><em>, </em><em>default='X_umap'</em>) – The name of the column in adata.obsm that will contain the
umap coordinates.</p></li>
<li><p><strong>associated_table</strong> (<em>str</em><em>, </em><em>optional</em>) – If set, use the corresponding key <cite>adata.obsm</cite> to calcuate the
UMAP. Takes priority over the layer argument.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>adata</strong> – Updated AnnData object with UMAP coordinates stored in the <cite>obsm</cite>
attribute. The key for the UMAP embedding in <cite>obsm</cite> is “X_umap” by
default.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>anndata.AnnData</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.transformations.run_utag_clustering">
<span class="sig-prename descclassname"><span class="pre">spac.transformations.</span></span><span class="sig-name descname"><span class="pre">run_utag_clustering</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">features</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">k</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">15</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">resolution</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_dist</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_pcs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">random_state</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">42</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_jobs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_iterations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slide_key</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Slide'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_annotation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'UTAG'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">associated_table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parallel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/transformations.html#run_utag_clustering"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.transformations.run_utag_clustering" title="Permalink to this definition"></a></dt>
<dd><p>Run UTAG clustering on the AnnData object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>anndata.AnnData</em>) – The AnnData object.</p></li>
<li><p><strong>features</strong> (<em>list</em>) – List of features to use for clustering or for PCA. Default
(None) is to use all.</p></li>
<li><p><strong>k</strong> (<em>int</em>) – The number of nearest neighbor to be used in creating the graph.
Default is 15.</p></li>
<li><p><strong>resolution</strong> (<em>float</em>) – Resolution parameter for the clustering, higher resolution produces
more clusters. Default is 1.</p></li>
<li><p><strong>max_dist</strong> (<em>float</em>) – Maximum distance to cut edges within a graph. Default is 20.</p></li>
<li><p><strong>n_principal_components</strong> (<em>int</em>) – Number of principal components to use for clustering.</p></li>
<li><p><strong>random_state</strong> (<em>int</em>) – Random state for reproducibility.</p></li>
<li><p><strong>n_jobs</strong> (<em>int</em>) – Number of jobs to run in parallel. Default is 5.</p></li>
<li><p><strong>n_iterations</strong> (<em>int</em>) – Number of iterations for the clustering.</p></li>
<li><p><strong>slide_key</strong> (<em>str</em>) – Key of adata.obs containing information on the batch structure
of the data.In general, for image data this will often be a variable
indicating the imageb so image-specific effects are removed from data.
Default is “Slide”.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>adata</strong> – Updated AnnData object with clustering results.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>anndata.AnnData</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.transformations.tsne">
<span class="sig-prename descclassname"><span class="pre">spac.transformations.</span></span><span class="sig-name descname"><span class="pre">tsne</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/transformations.html#tsne"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.transformations.tsne" title="Permalink to this definition"></a></dt>
<dd><p>Perform t-SNE transformation on specific layer information.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>anndata.AnnData</em>) – The AnnData object.</p></li>
<li><p><strong>layer</strong> (<em>str</em>) – Layer for phenograph cluster calculation.</p></li>
<li><p><strong>**kwargs</strong> – Parameters for scanpy.tl.tsne function.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>adata</strong> – Updated AnnData object with t-SNE coordinates.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>anndata.AnnData</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.transformations.z_score_normalization">
<span class="sig-prename descclassname"><span class="pre">spac.transformations.</span></span><span class="sig-name descname"><span class="pre">z_score_normalization</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_layer</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input_layer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/transformations.html#z_score_normalization"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.transformations.z_score_normalization" title="Permalink to this definition"></a></dt>
<dd><p>Compute z-scores for the provided AnnData object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>anndata.AnnData</em>) – The AnnData object containing the data to normalize.</p></li>
<li><p><strong>output_layer</strong> (<em>str</em>) – The name of the layer to store the computed z-scores.</p></li>
<li><p><strong>input_layer</strong> (<em>str</em><em>, </em><em>optional</em>) – The name of the layer in the AnnData object to normalize.
If None, the main data matrix .X is used.</p></li>
<li><p><strong>**kwargs</strong> (<em>dict</em><em>, </em><em>optional</em>) – Additional arguments to pass to scipy.stats.zscore.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<span class="target" id="module-spac.utag_functions"></span><dl class="py function">
<dt class="sig sig-object py" id="spac.utag_functions.add_probabilities_to_centroid">
<span class="sig-prename descclassname"><span class="pre">spac.utag_functions.</span></span><span class="sig-name descname"><span class="pre">add_probabilities_to_centroid</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name_to_output</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/utag_functions.html#add_probabilities_to_centroid"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.utag_functions.add_probabilities_to_centroid" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.utag_functions.custom_message_passing">
<span class="sig-prename descclassname"><span class="pre">spac.utag_functions.</span></span><span class="sig-name descname"><span class="pre">custom_message_passing</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'l1_norm'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/utag_functions.html#custom_message_passing"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.utag_functions.custom_message_passing" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.utag_functions.low_variance_filter">
<span class="sig-prename descclassname"><span class="pre">spac.utag_functions.</span></span><span class="sig-name descname"><span class="pre">low_variance_filter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/utag_functions.html#low_variance_filter"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.utag_functions.low_variance_filter" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.utag_functions.sparse_matrix_dstack">
<span class="sig-prename descclassname"><span class="pre">spac.utag_functions.</span></span><span class="sig-name descname"><span class="pre">sparse_matrix_dstack</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">matrices</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span><span class="p"><span class="pre">[</span></span><span class="pre">csr_matrix</span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">csr_matrix</span></span></span><a class="reference internal" href="_modules/spac/utag_functions.html#sparse_matrix_dstack"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.utag_functions.sparse_matrix_dstack" title="Permalink to this definition"></a></dt>
<dd><p>Diagonally stack sparse matrices.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.utag_functions.utag">
<span class="sig-prename descclassname"><span class="pre">spac.utag_functions.</span></span><span class="sig-name descname"><span class="pre">utag</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">channels_to_use</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slide_key</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Slide'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_key</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'UTAG</span> <span class="pre">Label'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_by_variance</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_dist</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">20.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">normalization_mode</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'l1_norm'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_spatial_connectivity</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_pcs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">apply_umap</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">umap_kwargs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Any</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">apply_clustering</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">clustering_method</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">['leiden']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">resolutions</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span><span class="p"><span class="pre">[</span></span><span class="pre">float</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">[0.05,</span> <span class="pre">0.1,</span> <span class="pre">0.3,</span> <span class="pre">1.0]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">leiden_kwargs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Any</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parc_kwargs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Any</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parallel</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">processes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">k</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">15</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">random_state</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">42</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/utag_functions.html#utag"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.utag_functions.utag" title="Permalink to this definition"></a></dt>
<dd><p>Discover tissue architechture in single-cell imaging data
by combining phenotypes and positional information of cells.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – AnnData object with spatial positioning of cells in obsm ‘spatial’ slot.</p></li>
<li><p><strong>channels_to_use</strong> (<em>Optional</em><em>[</em><em>Sequence</em><em>[</em><em>str</em><em>]</em><em>]</em>) – An optional sequence of strings used to subset variables to use.
Default (None) is to use all variables.</p></li>
<li><p><strong>max_dist</strong> (<em>float</em>) – Maximum distance to cut edges within a graph.
Should be adjusted depending on resolution of images.
For imaging mass cytometry, where resolution is 1um, 20 often
gives good results. Default is 20.</p></li>
<li><p><strong>slide_key</strong> (<em>{str</em><em>, </em><em>None}</em>) – Key of adata.obs containing information on the batch structure
of the data. In general, for image data this will often be a variable
indicating the image so image-specific effects are removed from data.
Default is “Slide”.</p></li>
<li><p><strong>save_key</strong> (<em>str</em>) – Key to be added to adata object holding the UTAG clusters.
Depending on the values of <cite>clustering_method</cite> and <cite>resolutions</cite>,
the final keys will be of the form: {save_key}_{method}_{resolution}”.
Default is “UTAG Label”.</p></li>
<li><p><strong>filter_by_variance</strong> (<em>bool</em>) – Whether to filter vairiables by variance.
Default is False, which keeps all variables.</p></li>
<li><p><strong>max_dist</strong> – Recommended values are between 20 to 50 depending on magnification.
Default is 20.</p></li>
<li><p><strong>normalization_mode</strong> (<em>str</em>) – Method to normalize adjacency matrix.
Default is “l1_norm”, any other value will not use normalization.</p></li>
<li><p><strong>keep_spatial_connectivity</strong> (<em>bool</em>) – Whether to keep sparse matrices of spatial connectivity and distance
in the obsp attribute of the resulting anndata object. This could be
useful in downstream applications.
Default is not to (False).</p></li>
<li><p><strong>n_pcs</strong> (<em>Number</em><em> of </em><em>principal components to use for clustering. Default is 10.</em>) – </p></li>
<li><p><strong>None</strong> (<em>If</em>) – </p></li>
<li><p><strong>features.</strong> (<em>no PCA is performed and clustering is done on</em>) – </p></li>
<li><p><strong>apply_umap</strong> (<em>bool</em>) – Whether to build a UMAP representation after message passing.
Default is False.</p></li>
<li><p><strong>umap_kwargs</strong> (<em>Dict</em><em>[</em><em>str</em><em>, </em><em>Any</em><em>]</em>) – Keyword arguments to be passed to scanpy.tl.umap for dimensionality
reduction after message passing. Default is 10.0.</p></li>
<li><p><strong>apply_clustering</strong> (<em>bool</em>) – Whether to cluster the message passed matrix.
Default is True.</p></li>
<li><p><strong>clustering_method</strong> (<em>Sequence</em><em>[</em><em>str</em><em>]</em>) – Which clustering method(s) to use for clustering of the
message passed matrix. Default is [“leiden”].</p></li>
<li><p><strong>resolutions</strong> (<em>Sequence</em><em>[</em><em>float</em><em>]</em>) – What resolutions should the methods in <cite>clustering_method</cite> be run at.
Default is [0.05, 0.1, 0.3, 1.0].</p></li>
<li><p><strong>leiden_kwargs</strong> (<em>dict</em><em>[</em><em>str</em><em>, </em><em>Any</em><em>]</em>) – Keyword arguments to pass to scanpy.tl.leiden.</p></li>
<li><p><strong>parc_kwargs</strong> (<em>dict</em><em>[</em><em>str</em><em>, </em><em>Any</em><em>]</em>) – Keyword arguments to pass to parc.PARC.</p></li>
<li><p><strong>parallel</strong> (<em>bool</em>) – Whether to run message passing part of algorithm in parallel.
Will accelerate the process but consume more memory.
Default is True.</p></li>
<li><p><strong>processes</strong> (<em>int</em>) – Number of processes to use in parallel.
Default is to use all available (-1).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>adata</strong> – AnnData object with UTAG domain predictions for each cell in adata.obs,
column <cite>save_key</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>AnnData</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.utag_functions.z_score">
<span class="sig-prename descclassname"><span class="pre">spac.utag_functions.</span></span><span class="sig-name descname"><span class="pre">z_score</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/utag_functions.html#z_score"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.utag_functions.z_score" title="Permalink to this definition"></a></dt>
<dd><p>Scale (divide by standard deviation) and center (subtract mean)
array-like objects.</p>
</dd></dl>

<span class="target" id="module-spac.utils"></span><dl class="py function">
<dt class="sig sig-object py" id="spac.utils.annotation_category_relations">
<span class="sig-prename descclassname"><span class="pre">spac.utils.</span></span><span class="sig-name descname"><span class="pre">annotation_category_relations</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source_annotation</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target_annotation</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/utils.html#annotation_category_relations"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.utils.annotation_category_relations" title="Permalink to this definition"></a></dt>
<dd><p>Calculates the count of unique relationships between two
annotations in an AnnData object.
Relationship is defined as a unique pair of values, one from the
‘source_annotation’ and one from the ‘target_annotation’.</p>
<p>Returns a DataFrame with columns ‘source_annotation’, ‘target_annotation’,
‘count’, ‘percentage_source’, and ‘percentage_target’.
Where ‘count’ represents the number of occurrences of each relationship,
percentage_source represents the percentage of the count of link
over the total count of the source label, and percentage_target represents
the percentage of the count of link over the total count of the target.</p>
<p>If the <cite>prefix</cite> is set to True, it appends “<a href="#id5"><span class="problematic" id="id6">source_</span></a>” and “<a href="#id7"><span class="problematic" id="id8">target_</span></a>”
prefixes to labels in the “source” and “target” columns, respectively.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – The annotated data matrix of shape <cite>n_obs</cite> * <cite>n_vars</cite>.
Rows correspond to cells and columns to genes.</p></li>
<li><p><strong>source_annotation</strong> (<em>str</em>) – The name of the source annotation column in the <cite>adata</cite> object.</p></li>
<li><p><strong>target_annotation</strong> (<em>str</em>) – The name of the target annotation column in the <cite>adata</cite> object.</p></li>
<li><p><strong>prefix</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, appends “<a href="#id9"><span class="problematic" id="id10">source_</span></a>” and “<a href="#id11"><span class="problematic" id="id12">target_</span></a>” prefixes to the
“source” and “target” columns, respectively.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>relationships</strong> – A DataFrame with the source and target categories,
their counts and their percentages.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.utils.check_annotation">
<span class="sig-prename descclassname"><span class="pre">spac.utils.</span></span><span class="sig-name descname"><span class="pre">check_annotation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parameter_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">should_exist</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/utils.html#check_annotation"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.utils.check_annotation" title="Permalink to this definition"></a></dt>
<dd><p>Perform common error checks for annotations in
anndata related objects.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>anndata.AnnData</em>) – The AnnData object to be checked.</p></li>
<li><p><strong>annotations</strong> (<em>str</em><em> or </em><em>list</em><em> of </em><em>str</em><em>, </em><em>optional</em>) – The annotation(s) to check for existence in adata.obs.</p></li>
<li><p><strong>should_exist</strong> (<em>bool</em><em>, </em><em>optional</em><em> (</em><em>default=True</em><em>)</em>) – Determines whether to check if elements exist in the
target list (True), or if they should not exist (False).</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>TypeError</strong> – If adata is not an instance of anndata.AnnData.</p></li>
<li><p><strong>ValueError</strong> – If any of the specified layers, annotations, or features do not exist.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.utils.check_column_name">
<span class="sig-prename descclassname"><span class="pre">spac.utils.</span></span><span class="sig-name descname"><span class="pre">check_column_name</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">column_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">field_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">symbol_checklist</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'!?,.'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/utils.html#check_column_name"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.utils.check_column_name" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.utils.check_distances">
<span class="sig-prename descclassname"><span class="pre">spac.utils.</span></span><span class="sig-name descname"><span class="pre">check_distances</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">distances</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/utils.html#check_distances"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.utils.check_distances" title="Permalink to this definition"></a></dt>
<dd><p>Check that the distances are valid: must be an array-like of
incremental positive values.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>distances</strong> (<em>list</em><em>, </em><em>tuple</em><em>, or </em><em>np.ndarray</em>) – The list of increasing distances for the neighborhood profile.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Raises a ValueError or TypeError if the distances are invalid.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>The distances must be a list of positive real numbers and must
be monotonically increasing.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.utils.check_feature">
<span class="sig-prename descclassname"><span class="pre">spac.utils.</span></span><span class="sig-name descname"><span class="pre">check_feature</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">features</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">should_exist</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/utils.html#check_feature"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.utils.check_feature" title="Permalink to this definition"></a></dt>
<dd><p>Perform common error checks for features in
anndata related objects.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>anndata.AnnData</em>) – The AnnData object to be checked.</p></li>
<li><p><strong>features</strong> (<em>str</em><em> or </em><em>list</em><em> of </em><em>str</em><em>, </em><em>optional</em>) – The feature(s) to check for existence in adata.var_names.</p></li>
<li><p><strong>should_exist</strong> (<em>bool</em><em>, </em><em>optional</em><em> (</em><em>default=True</em><em>)</em>) – Determines whether to check if elements exist in the
target list (True), or if they should not exist (False).</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>TypeError</strong> – If adata is not an instance of anndata.AnnData.</p></li>
<li><p><strong>ValueError</strong> – If any of the specified layers, annotations, or features do not exist.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.utils.check_label">
<span class="sig-prename descclassname"><span class="pre">spac.utils.</span></span><span class="sig-name descname"><span class="pre">check_label</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">should_exist</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">warning</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/utils.html#check_label"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.utils.check_label" title="Permalink to this definition"></a></dt>
<dd><p>Check if specified labels exist in a given annotation column in adata.obs.</p>
<p>This function verifies whether all or none of the specified labels
exist in the provided annotation column of an AnnData object. It ensures
that the input labels align with the expected categories present in
adata.obs[annotation].</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>anndata.AnnData</em>) – The AnnData object containing the annotation column.</p></li>
<li><p><strong>annotation</strong> (<em>str</em>) – The name of the annotation column in adata.obs to check against.</p></li>
<li><p><strong>labels</strong> (<em>str</em><em> or </em><em>list</em><em> of </em><em>str</em><em>, </em><em>optional</em>) – The label or list of labels to check for existence in the specified
annotation column. If None, no validation will be performed.</p></li>
<li><p><strong>should_exist</strong> (<em>bool</em><em>, </em><em>optional</em><em> (</em><em>default=True</em><em>)</em>) – Determines whether to check if elements exist in the target column
(True), or if they should not exist (False).</p></li>
<li><p><strong>warning</strong> (<em>bool</em><em>, </em><em>optional</em><em> (</em><em>default=False</em><em>)</em>) – If True, generate a warning instead of raising an exception if the
specified condition is not met.</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>TypeError</strong> – If adata is not an instance of anndata.AnnData.</p></li>
<li><p><strong>ValueError</strong> – If the specified annotation does not exist in adata.obs.
    If should_exist is True and any label does not exist in the
    annotation column.
    If should_exist is False and any label already exists in the
    annotation column.</p></li>
</ul>
</dd>
<dt class="field-odd">Warns<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>UserWarning</strong> – If the specified condition is not met and warning is True.</p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">check_label</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="s2">&quot;cell_type&quot;</span><span class="p">,</span> <span class="s2">&quot;B_cell&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">check_label</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">adata</span><span class="p">,</span> <span class="s2">&quot;cluster&quot;</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;Cluster1&quot;</span><span class="p">,</span> <span class="s2">&quot;Cluster2&quot;</span><span class="p">],</span> <span class="n">should_exist</span><span class="o">=</span><span class="kc">True</span>
<span class="gp">... </span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.utils.check_list_in_list">
<span class="sig-prename descclassname"><span class="pre">spac.utils.</span></span><span class="sig-name descname"><span class="pre">check_list_in_list</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input_type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">need_exist</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">warning</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/utils.html#check_list_in_list"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.utils.check_list_in_list" title="Permalink to this definition"></a></dt>
<dd><p>Check if items in a given list exist in a target list.</p>
<p>This function is used to validate whether all or none of the
items in a given list exist in a target list.
It helps to ensure that the input list contains only valid elements
that are present in the target list.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>input</strong> (<em>str</em><em> or </em><em>list</em><em> of </em><em>str</em><em> or </em><em>None</em>) – The input list or a single string element. If it is a string,
it will be converted to a list containing
only that string element. If <cite>None</cite>, no validation will be performed.</p></li>
<li><p><strong>input_name</strong> (<em>str</em>) – The name of the input list used for displaying helpful error messages.</p></li>
<li><p><strong>input_type</strong> (<em>str</em>) – The type of items in the input list
(e.g., “item”, “element”, “category”).</p></li>
<li><p><strong>target_list</strong> (<em>list</em><em> of </em><em>str</em>) – The target list containing valid items that the input list elements
should be compared against.</p></li>
<li><p><strong>need_exist</strong> (<em>bool</em><em>, </em><em>optional</em><em> (</em><em>default=True</em><em>)</em>) – <dl class="simple">
<dt>Determines whether to check if elements exist in the</dt><dd><p>target list (True), or if they should not exist (False).</p>
</dd>
<dt>warning: bool, optional (default=False)</dt><dd><p>If true, generate a warning instead of raising an exception</p>
</dd>
</dl>
</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If the <cite>input</cite> is not a string or a list of strings.
    If <cite>need_exist</cite> is True and any element of the input
        list does not exist in the target list.
    If <cite>need_exist</cite> is False and any element of the input
        list exists in the target list.</p>
</dd>
<dt class="field-odd">Warns<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>UserWarning</strong> – If the specified behavior is not present
and <cite>warning</cite> is True.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.utils.check_table">
<span class="sig-prename descclassname"><span class="pre">spac.utils.</span></span><span class="sig-name descname"><span class="pre">check_table</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tables</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">should_exist</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">associated_table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">warning</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/utils.html#check_table"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.utils.check_table" title="Permalink to this definition"></a></dt>
<dd><p>Perform common error checks for table (layers) or derived tables (obsm) in
anndata related objects.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>anndata.AnnData</em>) – The AnnData object to be checked.</p></li>
<li><p><strong>tables</strong> (<em>str</em><em> or </em><em>list</em><em> of </em><em>str</em><em>, </em><em>optional</em>) – The term “table” is equivalent to layer in anndata structure.
The layer(s) to check for existence in adata.layers.keys().</p></li>
<li><p><strong>should_exist</strong> (<em>bool</em><em>, </em><em>optional</em><em> (</em><em>default=True</em><em>)</em>) – Determines whether to check if elements exist in the
target list (True), or if they should not exist (False).</p></li>
<li><p><strong>associtated_table</strong> (<em>bool</em><em>, </em><em>optional</em><em> (</em><em>default=False</em><em>)</em>) – Determines whether to check if the passed tables names
should exist as layers or in obsm in the andata object.</p></li>
<li><p><strong>warning</strong> (<em>bool</em><em>, </em><em>optional</em><em> (</em><em>default=False</em><em>)</em>) – If True, generate a warning instead of raising an exception.</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>TypeError</strong> – If adata is not an instance of anndata.AnnData.</p></li>
<li><p><strong>ValueError</strong> – If any of the specified layers, annotations, obsm,
    or features do not exist.</p></li>
</ul>
</dd>
<dt class="field-odd">Warns<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>UserWarning</strong> – If any of the specified layers, annotations, obsm,
or features do not exist, and <cite>warning</cite> is True.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.utils.color_mapping">
<span class="sig-prename descclassname"><span class="pre">spac.utils.</span></span><span class="sig-name descname"><span class="pre">color_mapping</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">labels</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color_map</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'viridis'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">opacity</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rgba_mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_dict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/utils.html#color_mapping"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.utils.color_mapping" title="Permalink to this definition"></a></dt>
<dd><p>Map a list of labels to colors using a Matplotlib colormap and opacity.</p>
<p>This function assigns a unique color to each label in the provided list
using a specified colormap from Matplotlib. The generated colors can be
returned in either <cite>rgba</cite> or <cite>rgb</cite> format, suitable for visualization in
libraries like Plotly.</p>
<p>The function supports both continuous and discrete colormaps:
- Continuous colormaps interpolate smoothly between colors across a range.
- Discrete colormaps have a fixed number of distinct colors, and labels are</p>
<blockquote>
<div><p>distributed evenly across these colors.</p>
</div></blockquote>
<p>Opacity can be set with a value between 0 (fully transparent) and 1 (fully
opaque). The resulting colors are CSS-compatible strings.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>labels</strong> (<em>list</em>) – A list of unique labels to map to colors. The number of labels
determines how the colormap is sampled.</p></li>
<li><p><strong>color_map</strong> (<em>str</em><em>, </em><em>optional</em>) – The colormap name (e.g., ‘viridis’, ‘plasma’, ‘inferno’). It must be a
valid Matplotlib colormap. Default is ‘viridis’.</p></li>
<li><p><strong>opacity</strong> (<em>float</em><em>, </em><em>optional</em>) – Opacity (alpha channel) for colors, between 0 (transparent) and 1
(opaque). Default is 1.0.</p></li>
<li><p><strong>rgba_mode</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, returns colors in <cite>rgba</cite> format (e.g., <cite>rgba(255, 0, 0, 0.5)</cite>).
If False, returns <cite>rgb</cite> format (e.g., <cite>rgb(255, 0, 0)</cite>).
Default is True.</p></li>
<li><p><strong>return_dict</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, returns a dictionary where keys are labels, and values are the
corresponding colors. Default is False.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>label_colors</strong> – If <cite>return_dict</cite> is False, returns a list of color strings, one for
each label. If <cite>return_dict</cite> is True, returns a dictionary with label
keys and color values. The format of the colors depends on the
<cite>rgba_mode</cite> parameter.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[str] or dict</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – <ul class="simple">
<li><p>If <cite>opacity</cite> is not in the range [0, 1].
    - If <cite>color_map</cite> is not a valid Matplotlib colormap name.</p></li>
</ul>
</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Assign colors to labels with default settings:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">labels</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">color_mapping</span><span class="p">(</span><span class="n">labels</span><span class="p">)</span>
<span class="go">[&#39;rgba(68, 1, 84, 1.0)&#39;, &#39;rgba(58, 82, 139, 1.0)&#39;,</span>
<span class="go"> &#39;rgba(33, 145, 140, 1.0)&#39;]</span>
</pre></div>
</div>
<p>Use a different colormap with reduced opacity:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">color_mapping</span><span class="p">(</span><span class="n">labels</span><span class="p">,</span> <span class="n">color_map</span><span class="o">=</span><span class="s1">&#39;plasma&#39;</span><span class="p">,</span> <span class="n">opacity</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
<span class="go">[&#39;rgba(13, 8, 135, 0.5)&#39;, &#39;rgba(126, 3, 167, 0.5)&#39;,</span>
<span class="go"> &#39;rgba(240, 249, 33, 0.5)&#39;]</span>
</pre></div>
</div>
<p>Generate colors in <cite>rgb</cite> format:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">color_mapping</span><span class="p">(</span><span class="n">labels</span><span class="p">,</span> <span class="n">rgba_mode</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">[&#39;rgb(68, 1, 84)&#39;, &#39;rgb(58, 82, 139)&#39;, &#39;rgb(33, 145, 140)&#39;]</span>
</pre></div>
</div>
<p>Return a dictionary of labels and colors:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">color_mapping</span><span class="p">(</span><span class="n">labels</span><span class="p">,</span> <span class="n">return_dict</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">{&#39;A&#39;: &#39;rgba(68, 1, 84, 1.0)&#39;, &#39;B&#39;: &#39;rgba(58, 82, 139, 1.0)&#39;,</span>
<span class="go"> &#39;C&#39;: &#39;rgba(33, 145, 140, 1.0)&#39;}</span>
</pre></div>
</div>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>Continuous colormaps interpolate colors evenly based on the number of
labels.</p></li>
<li><p>Discrete colormaps divide labels evenly across available colors.</p></li>
<li><p>For more information on Matplotlib colormaps:
<a class="reference external" href="https://matplotlib.org/stable/users/explain/colors/colormaps.html">https://matplotlib.org/stable/users/explain/colors/colormaps.html</a></p></li>
</ul>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.utils.compute_boxplot_metrics">
<span class="sig-prename descclassname"><span class="pre">spac.utils.</span></span><span class="sig-name descname"><span class="pre">compute_boxplot_metrics</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">showfliers</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/utils.html#compute_boxplot_metrics"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.utils.compute_boxplot_metrics" title="Permalink to this definition"></a></dt>
<dd><p>Compute boxplot-related statistical metrics for a given dataset
efficiently.</p>
<dl class="simple">
<dt>Statistics include:</dt><dd><ul class="simple">
<li><p>Lower and upper whiskers (<cite>whislo</cite>, <cite>whishi</cite>),</p></li>
<li><p>First quartile (<cite>q1</cite>),</p></li>
<li><p>Median (<cite>median</cite>),</p></li>
<li><p>Third quartile (<cite>q3</cite>),</p></li>
<li><p>Mean (<cite>mean</cite>)</p></li>
<li><p>Outliers (<cite>fliers</cite>) [If <cite>showfliers</cite> is not None]</p></li>
</ul>
</dd>
</dl>
<p>It can identify outliers based on the ‘showfliers’ parameter, and
supports efficient handling of large datasets by downsampling outliers
when specified.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>pd.DataFrame</em>) – A pandas DataFrame containing the numerical data for which
the boxplot statistics are to be computed.</p></li>
<li><p><strong>annotation</strong> (<em>str</em><em>, </em><em>optional:</em>) – The annotation used to group the features</p></li>
<li><p><strong>showfliers</strong> (<em>{None</em><em>, </em><em>&quot;downsample&quot;</em><em>, </em><em>&quot;all&quot;}</em><em>, </em><em>default = None</em>) – Defines how outliers are handled
If ‘all’, all outliers are displayed in the boxplot.
If ‘downsample’, when num outliers is &gt;10k, they are downsampled to
10% of the original count.
If None, outliers are hidden.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>metrics</strong> – A dataframe with one row per feature/annotation grouping and
columns representing the calculated features</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.utils.get_defined_color_map">
<span class="sig-prename descclassname"><span class="pre">spac.utils.</span></span><span class="sig-name descname"><span class="pre">get_defined_color_map</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">defined_color_map</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">colorscale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'viridis'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/utils.html#get_defined_color_map"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.utils.get_defined_color_map" title="Permalink to this definition"></a></dt>
<dd><p>Retrieve or generate a predefined color mapping dictionary from an AnnData
object.</p>
<p>If <cite>defined_color_map</cite> is provided and found within <cite>adata.uns</cite>, the
corresponding dictionary is returned. Otherwise, if it is not provided, a
color mapping is generated using the unique values of the annotation column
specified by <cite>annotations</cite> and the given <cite>colorscale</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>anndata.AnnData</em>) – Annotated data matrix object that should contain a color mapping in its
<cite>uns</cite> attribute if a predefined mapping is desired.</p></li>
<li><p><strong>defined_color_map</strong> (<em>str</em><em>, </em><em>optional</em>) – The key in <cite>adata.uns</cite> that holds the predefined color mapping.
If None, a new mapping is generated using <cite>annotations</cite>.</p></li>
<li><p><strong>annotations</strong> (<em>str</em><em>, </em><em>optional</em>) – The annotation column name in adata.obs from which to obtain unique
labels if <cite>defined_color_map</cite> is not provided.</p></li>
<li><p><strong>colorscale</strong> (<em>str</em><em>, </em><em>optional</em>) – The Matplotlib colormap name to use when generating a color mapping if
<cite>defined_color_map</cite> is not provided. Default is ‘viridis’.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A dictionary mapping unique labels to colors.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>TypeError</strong> – If <cite>defined_color_map</cite> is provided but is not a string.</p></li>
<li><p><strong>ValueError</strong> – If a predefined mapping is requested but not found, or if neither
    <cite>defined_color_map</cite> nor <cite>annotations</cite> is provided.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.utils.regex_search_list">
<span class="sig-prename descclassname"><span class="pre">spac.utils.</span></span><span class="sig-name descname"><span class="pre">regex_search_list</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">regex_patterns</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">list_to_search</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/utils.html#regex_search_list"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.utils.regex_search_list" title="Permalink to this definition"></a></dt>
<dd><p>Perfrom regex (regular expression) search in a list and
return list of strings matching the regex pattern</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>regex_pattern</strong> (<em>str</em><em> or </em><em>a list</em><em> of </em><em>str</em>) – The regex pattern to look for,
single pattern or a list of patterns.</p></li>
<li><p><strong>list_to_search</strong> (<em>list</em><em> of </em><em>str</em>) – A list of string to seach for string that matches regex pattern.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A list of strings containing results from search.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list of str</p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">regex_pattern</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;^B.*&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">list_to_search</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;ABC&quot;</span><span class="p">,</span> <span class="s2">&quot;BC&quot;</span><span class="p">,</span> <span class="s2">&quot;AC&quot;</span><span class="p">,</span> <span class="s2">&quot;AB&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">result</span> <span class="o">=</span> <span class="n">regex_search_list</span><span class="p">(</span><span class="n">regex_pattern</span><span class="p">,</span> <span class="n">list_to_search</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">result</span><span class="p">)</span>
<span class="go">[&#39;BC&#39;]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.utils.spell_out_special_characters">
<span class="sig-prename descclassname"><span class="pre">spac.utils.</span></span><span class="sig-name descname"><span class="pre">spell_out_special_characters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">text</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/utils.html#spell_out_special_characters"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.utils.spell_out_special_characters" title="Permalink to this definition"></a></dt>
<dd><p>Convert special characters in a string to comply with NIDAP naming rules.</p>
<p>This function processes a string by replacing or removing disallowed
characters to ensure compatibility with NIDAP. Spaces, special symbols,
and certain substrings are replaced or transformed into readable and
regulation-compliant equivalents.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>text</strong> (<em>str</em>) – The input string to be processed and converted.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>str</em> – A sanitized string with special characters replaced or removed,
adhering to NIDAP naming conventions.</p></li>
<li><p><em>Processing Steps</em></p></li>
<li><p><em>—————-</em></p></li>
<li><ol class="arabic simple">
<li><p>Spaces are replaced with underscores (<cite>_</cite>).</p></li>
</ol>
</li>
<li><p><em>2. Substrings related to units (e.g., ‘µm²’) are replaced with text</em> – equivalents:
- ‘µm²’ -&gt; ‘um2’
- ‘µm’ -&gt; ‘um’</p></li>
<li><ol class="arabic simple" start="3">
<li><p>Hyphens (<cite>-</cite>) between letters are replaced with underscores (<cite>_</cite>).</p></li>
</ol>
</li>
<li><p><em>4. Certain special symbols are mapped to readable equivalents</em> –</p>
<ul>
<li><p><cite>+</cite> -&gt; <cite>_pos_</cite></p></li>
<li><p><cite>-</cite> -&gt; <cite>_neg_</cite></p></li>
<li><p><cite>&#64;</cite> -&gt; <cite>at</cite></p></li>
<li><p><cite>#</cite> -&gt; <cite>hash</cite></p></li>
<li><p><cite>&amp;</cite> -&gt; <cite>and</cite></p></li>
<li><p>And more (see Notes section for a full mapping).</p></li>
</ul>
</li>
<li><p><em>5. Remaining disallowed characters are removed (non-alphanumeric and</em> – non-underscore characters).</p></li>
<li><p><em>6. Consecutive underscores are consolidated into a single underscore.</em></p></li>
<li><p><em>7. Leading and trailing underscores are stripped.</em></p></li>
</ul>
</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>The following special character mappings are used:
- <cite>µ</cite> -&gt; <cite>u</cite>
- <cite>²</cite> -&gt; <cite>2</cite>
- <cite>/</cite> -&gt; <cite>slash</cite>
- <cite>=</cite> -&gt; <cite>equals</cite>
- <cite>!</cite> -&gt; <cite>exclamation</cite>
- <cite>|</cite> -&gt; <cite>pipe</cite>
- For a complete list, refer to the <cite>special_char_map</cite> in the code.</p>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">spell_out_special_characters</span><span class="p">(</span><span class="s2">&quot;Data µm²+Analysis #1-2&quot;</span><span class="p">)</span>
<span class="go">&#39;Data_um2_pos_Analysis_hash1_neg_2&#39;</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">spell_out_special_characters</span><span class="p">(</span><span class="s2">&quot;Invalid!Char@Format&quot;</span><span class="p">)</span>
<span class="go">&#39;Invalid_exclamation_Char_at_Format&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.utils.text_to_others">
<span class="sig-prename descclassname"><span class="pre">spac.utils.</span></span><span class="sig-name descname"><span class="pre">text_to_others</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">parameter</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">text</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'None'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">to_None</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">to_False</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">to_True</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">to_Int</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">to_Float</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/utils.html#text_to_others"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.utils.text_to_others" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<span class="target" id="module-spac.visualization"></span><dl class="py function">
<dt class="sig sig-object py" id="spac.visualization.boxplot">
<span class="sig-prename descclassname"><span class="pre">spac.visualization.</span></span><span class="sig-name descname"><span class="pre">boxplot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">second_annotation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">features</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log_scale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/visualization.html#boxplot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.visualization.boxplot" title="Permalink to this definition"></a></dt>
<dd><p>Create a boxplot visualization of the features in the passed adata object.
This function offers flexibility in how the boxplots are displayed,
based on the arguments provided.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>anndata.AnnData</em>) – The AnnData object.</p></li>
<li><p><strong>annotation</strong> (<em>str</em><em>, </em><em>optional</em>) – Annotation to determine if separate plots are needed for every label.</p></li>
<li><p><strong>second_annotation</strong> (<em>str</em><em>, </em><em>optional</em>) – Second annotation to further divide the data.</p></li>
<li><p><strong>layer</strong> (<em>str</em><em>, </em><em>optional</em>) – The name of the matrix layer to use. If not provided,
uses the main data matrix adata.X.</p></li>
<li><p><strong>ax</strong> (<em>matplotlib.axes.Axes</em><em>, </em><em>optional</em>) – An existing Axes object to draw the plot onto, optional.</p></li>
<li><p><strong>features</strong> (<em>list</em><em>, </em><em>optional</em>) – List of feature names to be plotted.
If not provided, all features will be plotted.</p></li>
<li><p><strong>log_scale</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, the Y-axis will be in log scale. Default is False.</p></li>
<li><p><strong>**kwargs</strong> – <p>Additional arguments to pass to seaborn.boxplot.
Key arguments include:
- <cite>orient</cite>: Determines the orientation of the plot.
* “v”: Vertical orientation (default). In this case, categorical data</p>
<blockquote>
<div><p>will be plotted on the x-axis, and the boxplots will be vertical.</p>
</div></blockquote>
<ul>
<li><dl class="simple">
<dt>”h”: Horizontal orientation. Categorical data will be plotted on the</dt><dd><p>y-axis, and the boxplots will be horizontal.</p>
</dd>
</dl>
</li>
</ul>
</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>fig, ax</strong> – The created figure and axes for the plot.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>matplotlib.figure.Figure, matplotlib.axes.Axes</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<ul class="simple">
<li><p>Multiple features boxplot: boxplot(adata, features=[‘GeneA’,’GeneB’])</p></li>
<li><p>Boxplot grouped by a single annotation:
boxplot(adata, features=[‘GeneA’], annotation=’cell_type’)</p></li>
<li><p>Boxplot for multiple features grouped by a single annotation:
boxplot(adata, features=[‘GeneA’, ‘GeneB’], annotation=’cell_type’)</p></li>
<li><p>Nested grouping by two annotations: boxplot(adata, features=[‘GeneA’],
annotation=’cell_type’, second_annotation=’treatment’)</p></li>
</ul>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.visualization.boxplot_interactive">
<span class="sig-prename descclassname"><span class="pre">spac.visualization.</span></span><span class="sig-name descname"><span class="pre">boxplot_interactive</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">features</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">showfliers</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log_scale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">orient</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'v'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figure_width</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3.2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figure_height</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figure_dpi</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">200</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">defined_color_map</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation_colorscale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'viridis'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">feature_colorscale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'seismic'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figure_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'interactive'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_metrics</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/visualization.html#boxplot_interactive"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.visualization.boxplot_interactive" title="Permalink to this definition"></a></dt>
<dd><p>Generate a boxplot for given features from an AnnData object.</p>
<p>This function visualizes the distribution of gene expression
(or other features) across different annotations in the provided data.
It can handle various options such as log-transformation, feature
selection, and handling of outliers.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – An AnnData object containing the data to plot. The expression matrix
is accessed via <cite>adata.X</cite> or <cite>adata.layers[layer]</cite>, and annotations
are taken from <cite>adata.obs</cite>.</p></li>
<li><p><strong>annotation</strong> (<em>str</em><em>, </em><em>optional</em>) – The name of the annotation column (e.g., cell type or sample
condition) from <cite>adata.obs</cite> used to group the features. If <cite>None</cite>, no
grouping is applied.</p></li>
<li><p><strong>layer</strong> (<em>str</em><em>, </em><em>optional</em>) – The name of the layer from <cite>adata.layers</cite> to use. If <cite>None</cite>, <cite>adata.X</cite>
is used.</p></li>
<li><p><strong>ax</strong> (<em>matplotlib.axes.Axes</em><em>, </em><em>optional</em>) – The figure to plot the boxplot onto. If <cite>None</cite>, a new figure is
created.</p></li>
<li><p><strong>features</strong> (<em>list</em><em> of </em><em>str</em><em>, </em><em>optional</em>) – The list of features (genes) to plot. If <cite>None</cite>, all features are
included.</p></li>
<li><p><strong>showfliers</strong> (<em>{None</em><em>, </em><em>&quot;downsample&quot;</em><em>, </em><em>&quot;all&quot;}</em><em>, </em><em>default = None</em>) – If ‘all’, all outliers are displayed in the boxplot.
If ‘downsample’, when num outliers is &gt;10k, they are downsampled to
10% of the original count.
If None, outliers are hidden.</p></li>
<li><p><strong>log_scale</strong> (<em>bool</em><em>, </em><em>default=False</em>) – If True, the log1p transformation is applied to the features before
plotting. This option is disabled if negative values are found in the
features.</p></li>
<li><p><strong>orient</strong> (<em>{&quot;v&quot;</em><em>, </em><em>&quot;h&quot;}</em><em>, </em><em>default=&quot;v&quot;</em>) – The orientation of the boxplots: “v” for vertical, “h” for horizontal.</p></li>
<li><p><strong>figure_width</strong> (<em>int</em><em>, </em><em>optional</em>) – Width of the figure in inches. Default is 3.2.</p></li>
<li><p><strong>figure_height</strong> (<em>int</em><em>, </em><em>optional</em>) – Height of the figure in inches. Default is 2.</p></li>
<li><p><strong>figure_dpi</strong> (<em>int</em><em>, </em><em>optional</em>) – DPI (dots per inch) for the figure. Default is 200.</p></li>
<li><p><strong>defined_color_map</strong> (<em>str</em><em>, </em><em>optional</em>) – Key in ‘adata.uns’ holding a pre-computed color dictionary.
Falls back to automatic generation from ‘annotation’ values.</p></li>
<li><p><strong>ax</strong> – A Matplotlib Axes object. Currently, this parameter is not used by the
underlying plotting functions (Seaborn’s <cite>catplot</cite>/<cite>displot</cite>), which
will always generate a new figure and axes. The <cite>ax</cite> key in the
returned dictionary will contain the Axes from these new plots.
This parameter is maintained for API consistency and potential
future enhancements. Default is None.</p></li>
<li><p><strong>**kwargs</strong> (<em>dict</em>) – Additional arguments for seaborn figure-level functions.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl>
<dt>fig<span class="classifier">plotly.graph_objects.Figure or str</span></dt><dd><dl>
<dt>The generated boxplot figure, which can be either:</dt><dd><ul class="simple">
<li><p>If <cite>figure_type</cite> is “static”: A base64-encoded PNG</p></li>
</ul>
<p>image string
- If <cite>figure_type</cite> is “interactive”: A Plotly figure object</p>
</dd>
</dl>
</dd>
<dt>df<span class="classifier">pd.DataFrame</span></dt><dd><p>A DataFrame containing the features and their corresponding values.</p>
</dd>
<dt>metrics<span class="classifier">pd.DataFrame</span></dt><dd><p>A DataFrame containing the computed boxplot metrics (if
<cite>return_metrics</cite> is True).</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>A dictionary containing the following keys</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.visualization.dimensionality_reduction_plot">
<span class="sig-prename descclassname"><span class="pre">spac.visualization.</span></span><span class="sig-name descname"><span class="pre">dimensionality_reduction_plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">feature</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">associated_table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/visualization.html#dimensionality_reduction_plot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.visualization.dimensionality_reduction_plot" title="Permalink to this definition"></a></dt>
<dd><p>Visualize scatter plot in PCA, t-SNE, UMAP, or associated table.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>anndata.AnnData</em>) – The AnnData object with coordinates precomputed by the ‘tsne’ or ‘UMAP’
function and stored in ‘adata.obsm[“X_tsne”]’ or ‘adata.obsm[“X_umap”]’</p></li>
<li><p><strong>method</strong> (<em>str</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – Dimensionality reduction method to visualize.
Choose from {‘tsne’, ‘umap’, ‘pca’}.</p></li>
<li><p><strong>annotation</strong> (<em>str</em><em>, </em><em>optional</em>) – The name of the column in <cite>adata.obs</cite> to use for coloring
the scatter plot points based on cell annotations.
Takes precedence over <cite>feature</cite>.</p></li>
<li><p><strong>feature</strong> (<em>str</em><em>, </em><em>optional</em>) – The name of the gene or feature in <cite>adata.var_names</cite> to use
for coloring the scatter plot points based on feature expression.</p></li>
<li><p><strong>layer</strong> (<em>str</em><em>, </em><em>optional</em>) – The name of the data layer in <cite>adata.layers</cite> to use for visualization.
If None, the main data matrix <cite>adata.X</cite> is used.</p></li>
<li><p><strong>ax</strong> (<em>matplotlib.axes.Axes</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – A matplotlib axes object to plot on.
If not provided, a new figure and axes will be created.</p></li>
<li><p><strong>associated_table</strong> (<em>str</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – Name of the key in <cite>obsm</cite> that contains the numpy array. Takes
precedence over <cite>method</cite></p></li>
<li><p><strong>**kwargs</strong> – Parameters passed to visualize_2D_scatter function,
including point_size.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>fig</strong> (<em>matplotlib.figure.Figure</em>) – The created figure for the plot.</p></li>
<li><p><strong>ax</strong> (<em>matplotlib.axes.Axes</em>) – The axes of the plot.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.visualization.heatmap">
<span class="sig-prename descclassname"><span class="pre">spac.visualization.</span></span><span class="sig-name descname"><span class="pre">heatmap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">column</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/visualization.html#heatmap"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.visualization.heatmap" title="Permalink to this definition"></a></dt>
<dd><p>Plot the heatmap of the mean feature of cells that belong to a <cite>column</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>anndata.AnnData</em>) – The AnnData object.</p></li>
<li><p><strong>column</strong> (<em>str</em>) – Name of member of adata.obs to plot the histogram.</p></li>
<li><p><strong>layer</strong> (<em>str</em><em>, </em><em>default None</em>) – The name of the <cite>adata</cite> layer to use to calculate the mean feature.</p></li>
<li><p><strong>**kwargs</strong> – Parameters passed to seaborn heatmap function.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>pandas.DataFrame</em> – A dataframe tha has the labels as indexes the mean feature for every
marker.</p></li>
<li><p><em>matplotlib.figure.Figure</em> – The figure of the heatmap.</p></li>
<li><p><em>matplotlib.axes._subplots.AxesSubplot</em> – The AsxesSubplot of the heatmap.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.visualization.hierarchical_heatmap">
<span class="sig-prename descclassname"><span class="pre">spac.visualization.</span></span><span class="sig-name descname"><span class="pre">hierarchical_heatmap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">features</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cluster_feature</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cluster_annotations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">standard_scale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">z_score</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'annotation'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">swap_axes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rotate_label</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/visualization.html#hierarchical_heatmap"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.visualization.hierarchical_heatmap" title="Permalink to this definition"></a></dt>
<dd><p>Generates a hierarchical clustering heatmap and dendrogram.
By default, the dataset is assumed to have features as columns and
annotations as rows. Cells are grouped by annotation (e.g., phenotype),
and for each group, the average expression intensity of each feature
(e.g., protein or marker) is computed. The heatmap is plotted using
seaborn’s clustermap.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>anndata.AnnData</em>) – The AnnData object.</p></li>
<li><p><strong>annotation</strong> (<em>str</em>) – Name of the annotation in adata.obs to group by and calculate mean
intensity.</p></li>
<li><p><strong>features</strong> (<em>list</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – List of feature names (e.g., markers) to be included in the
visualization. If None, all features are used. Default is None.</p></li>
<li><p><strong>layer</strong> (<em>str</em><em>, </em><em>optional</em>) – The name of the <cite>adata</cite> layer to use to calculate the mean intensity.
If not provided, uses the main matrix. Default is None.</p></li>
<li><p><strong>cluster_feature</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, perform hierarchical clustering on the feature axis.
Default is False.</p></li>
<li><p><strong>cluster_annotations</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, perform hierarchical clustering on the annotations axis.
Default is False.</p></li>
<li><p><strong>standard_scale</strong> (<em>int</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – Whether to standard scale data (0: row-wise or 1: column-wise).
Default is None.</p></li>
<li><p><strong>z_score</strong> (<em>str</em><em>, </em><em>optional</em>) – Specifies the axis for z-score normalization. Can be “feature” or
“annotation”. Default is “annotation”.</p></li>
<li><p><strong>swap_axes</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, switches the axes of the heatmap, effectively transposing
the dataset. By default (when False), annotations are on the vertical
axis (rows) and features are on the horizontal axis (columns).
When set to True, features will be on the vertical axis and
annotations on the horizontal axis. Default is False.</p></li>
<li><p><strong>rotate_label</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, rotate x-axis labels by 45 degrees. Default is False.</p></li>
<li><p><strong>**kwargs</strong> – <p>Additional parameters passed to <cite>sns.clustermap</cite> function or its
underlying functions. Some essential parameters include:
- <cite>cmap</cite> : colormap</p>
<blockquote>
<div><p>Colormap to use for the heatmap. It’s an argument for the underlying
<cite>sns.heatmap()</cite> used within <cite>sns.clustermap()</cite>. Examples include
“viridis”, “plasma”, “coolwarm”, etc.</p>
</div></blockquote>
<ul>
<li><p><cite>{row,col}_colors</cite> : Lists or DataFrames
Colors to use for annotating the rows/columns. Useful for visualizing
additional categorical information alongside the main heatmap.</p></li>
<li><p><cite>{dendrogram,colors}_ratio</cite> : tuple(float)
Control the size proportions of the dendrogram and the color labels
relative to the main heatmap.</p></li>
<li><p><cite>cbar_pos</cite> : tuple(float) or None
Specify the position and size of the colorbar in the figure. If set
to None, no colorbar will be added.</p></li>
<li><p><cite>tree_kws</cite> : dict
Customize the appearance of the dendrogram tree. Passes additional
keyword arguments to the underlying
<cite>matplotlib.collections.LineCollection</cite>.</p></li>
<li><p><cite>method</cite> : str
The linkage algorithm to use for the hierarchical clustering.
Defaults to ‘centroid’ in the function, but can be changed.</p></li>
<li><p><cite>metric</cite> : str
The distance metric to use for the hierarchy. Defaults to ‘euclidean’
in the function.</p></li>
</ul>
</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>mean_intensity</strong> (<em>pandas.DataFrame</em>) – A DataFrame containing the mean intensity of cells for each annotation.</p></li>
<li><p><strong>clustergrid</strong> (<em>seaborn.matrix.ClusterGrid</em>) – The seaborn ClusterGrid object representing the heatmap and
dendrograms.</p></li>
<li><p><strong>dendrogram_data</strong> (<em>dict</em>) – A dictionary containing hierarchical clustering linkage data for both
rows and columns. These linkage matrices can be used to generate
dendrograms with tools like scipy’s dendrogram function. This offers
flexibility in customizing and plotting dendrograms as needed.</p></li>
</ul>
</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>import matplotlib.pyplot as plt
import pandas as pd
import anndata
from spac.visualization import hierarchical_heatmap
X = pd.DataFrame([[1, 2], [3, 4]], columns=[‘gene1’, ‘gene2’])
annotation = pd.DataFrame([‘type1’, ‘type2’], columns=[‘cell_type’])
all_data = anndata.AnnData(X=X, obs=annotation)</p>
<dl class="simple">
<dt>mean_intensity, clustergrid, dendrogram_data = hierarchical_heatmap(</dt><dd><p>all_data,
“cell_type”,
layer=None,
z_score=”annotation”,
swap_axes=True,
cluster_feature=False,
cluster_annotations=True</p>
</dd>
</dl>
<p>)</p>
<p># To display a standalone dendrogram using the returned linkage matrix:
import scipy.cluster.hierarchy as sch
import numpy as np
import matplotlib.pyplot as plt</p>
<p># Convert the linkage data to type double
dendro_col_data = np.array(dendrogram_data[‘col_linkage’], dtype=np.double)</p>
<p># Ensure the linkage matrix has at least two dimensions and
more than one linkage
if dendro_col_data.ndim == 2 and dendro_col_data.shape[0] &gt; 1:</p>
<blockquote>
<div><p>fig, ax = plt.subplots(figsize=(10, 7))
sch.dendrogram(dendro_col_data, ax=ax)
plt.title(‘Standalone Col Dendrogram’)
plt.show()</p>
</div></blockquote>
<dl class="simple">
<dt>else:</dt><dd><p>print(“Insufficient data to plot a dendrogram.”)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.visualization.histogram">
<span class="sig-prename descclassname"><span class="pre">spac.visualization.</span></span><span class="sig-name descname"><span class="pre">histogram</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">feature</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_by</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">together</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x_log_scale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y_log_scale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/visualization.html#histogram"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.visualization.histogram" title="Permalink to this definition"></a></dt>
<dd><p>Plot the histogram of cells based on a specific feature from adata.X
or annotation from adata.obs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>anndata.AnnData</em>) – The AnnData object.</p></li>
<li><p><strong>feature</strong> (<em>str</em><em>, </em><em>optional</em>) – Name of continuous feature from adata.X to plot its histogram.</p></li>
<li><p><strong>annotation</strong> (<em>str</em><em>, </em><em>optional</em>) – Name of the annotation from adata.obs to plot its histogram.</p></li>
<li><p><strong>layer</strong> (<em>str</em><em>, </em><em>optional</em>) – Name of the layer in adata.layers to plot its histogram.</p></li>
<li><p><strong>group_by</strong> (<em>str</em><em>, </em><em>default None</em>) – Choose either to group the histogram by another column.</p></li>
<li><p><strong>together</strong> (<em>bool</em><em>, </em><em>default False</em>) – If True, and if group_by != None, create one plot combining all groups.
If False, create separate histograms for each group.
The appearance of combined histograms can be controlled using the
<cite>multiple</cite> and <cite>element</cite> parameters in <a href="#id1"><span class="problematic" id="id2">**</span></a>kwargs.
To control how the histograms are normalized (e.g., to divide the
histogram by the number of elements in every group), use the <cite>stat</cite>
parameter in <a href="#id3"><span class="problematic" id="id4">**</span></a>kwargs. For example, set <cite>stat=”probability”</cite> to show
the relative frequencies of each group.</p></li>
<li><p><strong>ax</strong> (<em>matplotlib.axes.Axes</em><em>, </em><em>optional</em>) – An existing Axes object to draw the plot onto, optional.</p></li>
<li><p><strong>x_log_scale</strong> (<em>bool</em><em>, </em><em>default False</em>) – If True, the data will be transformed using np.log1p before plotting,
and the x-axis label will be adjusted accordingly.</p></li>
<li><p><strong>y_log_scale</strong> (<em>bool</em><em>, </em><em>default False</em>) – If True, the y-axis will be set to log scale.</p></li>
<li><p><strong>**kwargs</strong> – <p>Additional keyword arguments passed to seaborn histplot function.
Key arguments include:
- <cite>multiple</cite>: Determines how the subsets of data are displayed</p>
<blockquote>
<div><dl class="simple">
<dt>on the same axes. Options include:</dt><dd><ul>
<li><dl class="simple">
<dt>”layer”: Draws each subset on top of the other</dt><dd><p>without adjustments.</p>
</dd>
</dl>
</li>
<li><p>”dodge”: Dodges bars for each subset side by side.</p></li>
<li><p>”stack”: Stacks bars for each subset on top of each other.</p></li>
<li><p>”fill”: Adjusts bar heights to fill the axes.</p></li>
</ul>
</dd>
</dl>
</div></blockquote>
<ul>
<li><dl class="simple">
<dt><cite>element</cite>: Determines the visual representation of the bins.</dt><dd><dl class="simple">
<dt>Options include:</dt><dd><ul>
<li><p>”bars”: Displays the typical bar-style histogram (default).</p></li>
<li><p>”step”: Creates a step line plot without bars.</p></li>
<li><dl class="simple">
<dt>”poly”: Creates a polygon where the bottom edge represents</dt><dd><p>the x-axis and the top edge the histogram’s bins.</p>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><cite>log_scale</cite>: Determines if the data should be plotted on</dt><dd><p>a logarithmic scale.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><cite>stat</cite>: Determines the statistical transformation to use on the data</dt><dd><dl class="simple">
<dt>for the histogram. Options include:</dt><dd><ul>
<li><p>”count”: Show the counts of observations in each bin.</p></li>
<li><p>”frequency”: Show the number of observations divided
by the bin width.</p></li>
<li><dl class="simple">
<dt>”density”: Normalize such that the total area of the histogram</dt><dd><p>equals 1.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>”probability”: Normalize such that each bar’s height reflects</dt><dd><p>the probability of observing that bin.</p>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><cite>bins</cite>: Specification of hist bins.</dt><dd><p>Can be a number (indicating the number of bins) or a list
(indicating bin edges). For example, <cite>bins=10</cite> will create 10 bins,
while <cite>bins=[0, 1, 2, 3]</cite> will create bins [0,1), [1,2), [2,3].
If not provided, the binning will be determined automatically.
Note, don’t pass a numpy array, only python lists or strs/numbers.</p>
</dd>
</dl>
</li>
</ul>
</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>fig<span class="classifier">matplotlib.figure.Figure</span></dt><dd><p>The created figure for the plot.</p>
</dd>
<dt>axs<span class="classifier">matplotlib.axes.Axes or list of Axes</span></dt><dd><p>The Axes object(s) of the histogram plot(s). Returns a single Axes
if only one plot is created, otherwise returns a list of Axes.</p>
</dd>
<dt>df<span class="classifier">pandas.DataFrame</span></dt><dd><p>DataFrame containing the data used for plotting the histogram.</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>A dictionary containing the following</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.visualization.interactive_spatial_plot">
<span class="sig-prename descclassname"><span class="pre">spac.visualization.</span></span><span class="sig-name descname"><span class="pre">interactive_spatial_plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">feature</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dot_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dot_transparency</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.75</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation_colorscale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'rainbow'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">feature_colorscale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'balance'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figure_width</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">6</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figure_height</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">4</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figure_dpi</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">200</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">font_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">12</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stratify_by</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">defined_color_map</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reverse_y_axis</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmin</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/visualization.html#interactive_spatial_plot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.visualization.interactive_spatial_plot" title="Permalink to this definition"></a></dt>
<dd><p>Create an interactive scatter plot for
spatial data using provided annotations.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – Annotated data matrix object,
must have a .obsm attribute with ‘spatial’ key.</p></li>
<li><p><strong>annotations</strong> (<em>list</em><em> of </em><em>str</em><em> or </em><em>str</em><em>, </em><em>optional</em>) – Column(s) in <cite>adata.obs</cite> that contain the annotations to plot.
If a single string is provided, it will be converted to a list.
The interactive plot will show all the labels in the annotation
columns passed.</p></li>
<li><p><strong>feature</strong> (<em>str</em><em>, </em><em>optional</em>) – If annotation is None, the name of the gene or feature
in <cite>adata.var_names</cite> to use for coloring the scatter plot points
based on feature expression.</p></li>
<li><p><strong>layer</strong> (<em>str</em><em>, </em><em>optional</em>) – If feature is not None, the name of the data layer in <cite>adata.layers</cite>
to use for visualization. If None, the main data matrix <cite>adata.X</cite> is
used.</p></li>
<li><p><strong>dot_size</strong> (<em>float</em><em>, </em><em>optional</em>) – Size of the scatter dots in the plot. Default is 1.5.</p></li>
<li><p><strong>dot_transparency</strong> (<em>float</em><em>, </em><em>optional</em>) – Transparancy level of the scatter dots. Default is 0.75.</p></li>
<li><p><strong>annotation_colorscale</strong> (<em>str</em><em>, </em><em>optional</em>) – Name of the color scale to use for the dots when annotation
is used. Default is ‘Viridis’.</p></li>
<li><p><strong>feature_colorscale</strong> (<em>srt</em><em>, </em><em>optional</em>) – Name of the color scale to use for the dots when feature
is used. Default is ‘seismic’.</p></li>
<li><p><strong>figure_width</strong> (<em>int</em><em>, </em><em>optional</em>) – Width of the figure in inches. Default is 12.</p></li>
<li><p><strong>figure_height</strong> (<em>int</em><em>, </em><em>optional</em>) – Height of the figure in inches. Default is 8.</p></li>
<li><p><strong>figure_dpi</strong> (<em>int</em><em>, </em><em>optional</em>) – DPI (dots per inch) for the figure. Default is 200.</p></li>
<li><p><strong>font_size</strong> (<em>int</em><em>, </em><em>optional</em>) – Font size for text in the plot. Default is 12.</p></li>
<li><p><strong>stratify_by</strong> (<em>str</em><em>, </em><em>optional</em>) – Column in <cite>adata.obs</cite> to stratify the plot. Default is None.</p></li>
<li><p><strong>defined_color_map</strong> (<em>str</em><em>, </em><em>optional</em>) – Predefined color mapping stored in adata.uns for specific labels.
Default is None, which will generate the color mapping automatically.</p></li>
<li><p><strong>reverse_y_axis</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, reverse the Y-axis of the plot. Default is False.</p></li>
<li><p><strong>cmin</strong> (<em>float</em><em>, </em><em>optional</em>) – Minimum value for the color scale when using features.
Default is None.</p></li>
<li><p><strong>cmax</strong> (<em>float</em><em>, </em><em>optional</em>) – Maximum value for the color scale when using features.
Default is None.</p></li>
<li><p><strong>**kwargs</strong> – Additional keyword arguments for customization.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A list of dictionaries, each containing the following keys:
- “image_name”: str, the name of the generated image.
- “image_object”: Plotly Figure object.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list of dict</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>This function is tailored for spatial single-cell data and expects the
AnnData object to have spatial coordinates in its <cite>.obsm</cite> attribute under
the ‘spatial’ key.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.visualization.plot_ripley_l">
<span class="sig-prename descclassname"><span class="pre">spac.visualization.</span></span><span class="sig-name descname"><span class="pre">plot_ripley_l</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">phenotypes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">regions</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sims</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_df</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/visualization.html#plot_ripley_l"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.visualization.plot_ripley_l" title="Permalink to this definition"></a></dt>
<dd><p>Plot Ripley’s L statistic for multiple bins and different regions
for a given pair of phenotypes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – AnnData object containing Ripley’s L results in <cite>adata.uns[‘ripley_l’]</cite>.</p></li>
<li><p><strong>phenotypes</strong> (<em>tuple</em><em> of </em><em>str</em>) – A tuple of two phenotypes: (center_phenotype, neighbor_phenotype).</p></li>
<li><p><strong>regions</strong> (<em>list</em><em> of </em><em>str</em><em>, </em><em>optional</em>) – A list of region labels to plot. If None, plot all available regions.
Default is None.</p></li>
<li><p><strong>sims</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to plot the simulation results. Default is False.</p></li>
<li><p><strong>return_df</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to return the DataFrame containing the Ripley’s L results.</p></li>
<li><p><strong>kwargs</strong> (<em>dict</em><em>, </em><em>optional</em>) – Additional keyword arguments to pass to <cite>seaborn.lineplot</cite>.</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If the Ripley L results are not found in <cite>adata.uns[‘ripley_l’]</cite>.</p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><ul class="simple">
<li><p><strong>ax</strong> (<em>matplotlib.axes.Axes</em>) – The Axes object containing the plot, which can be further modified.</p></li>
<li><p><strong>df</strong> (<em>pandas.DataFrame, optional</em>) – The DataFrame containing the Ripley’s L results, if <cite>return_df</cite> is True.</p></li>
</ul>
</p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ax</span> <span class="o">=</span> <span class="n">plot_ripley_l</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">adata</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">phenotypes</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;Phenotype1&#39;</span><span class="p">,</span> <span class="s1">&#39;Phenotype2&#39;</span><span class="p">),</span>
<span class="gp">... </span>    <span class="n">regions</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;region1&#39;</span><span class="p">,</span> <span class="s1">&#39;region2&#39;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
<p>This returns the <cite>Axes</cite> object for further customization and displays the plot.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.visualization.present_summary_as_figure">
<span class="sig-prename descclassname"><span class="pre">spac.visualization.</span></span><span class="sig-name descname"><span class="pre">present_summary_as_figure</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">summary_dict</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Figure</span></span></span><a class="reference internal" href="_modules/spac/visualization.html#present_summary_as_figure"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.visualization.present_summary_as_figure" title="Permalink to this definition"></a></dt>
<dd><p>Build a static Plotly figure (using a table) to depict the
summary dictionary.</p>
<dl class="simple">
<dt>The figure includes columns:</dt><dd><ul class="simple">
<li><p>Column name</p></li>
<li><p>Data type</p></li>
<li><p>Count of missing values</p></li>
<li><p>Missing indices (as a string)</p></li>
<li><p>Summary details (formatted as JSON for readability)</p></li>
</ul>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>summary_dict</strong> (<em>dict</em>) – The summary dictionary returned from summarize_dataframe.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A static Plotly table figure representing the summary.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>plotly.graph_objects.Figure</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.visualization.present_summary_as_html">
<span class="sig-prename descclassname"><span class="pre">spac.visualization.</span></span><span class="sig-name descname"><span class="pre">present_summary_as_html</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">summary_dict</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">str</span></span></span><a class="reference internal" href="_modules/spac/visualization.html#present_summary_as_html"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.visualization.present_summary_as_html" title="Permalink to this definition"></a></dt>
<dd><p>Build an HTML string that presents the summary information
intuitively.</p>
<dl class="simple">
<dt>For each specified column, the HTML includes:</dt><dd><ul class="simple">
<li><p>Column name and data type</p></li>
<li><p>Count and list of missing indices</p></li>
<li><p>Summary details presented in a table (for numeric: stats;
categorical: unique values and counts)</p></li>
</ul>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>summary_dict</strong> (<em>dict</em>) – The summary dictionary returned by summarize_dataframe.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>HTML string representing the summary.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.visualization.relational_heatmap">
<span class="sig-prename descclassname"><span class="pre">spac.visualization.</span></span><span class="sig-name descname"><span class="pre">relational_heatmap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source_annotation</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target_annotation</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color_map</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'mint'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/visualization.html#relational_heatmap"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.visualization.relational_heatmap" title="Permalink to this definition"></a></dt>
<dd><p>Generates a relational heatmap from the given AnnData object.
The color map refers to matplotlib color maps, default is mint.
For more information on colormaps, see:
<a class="reference external" href="https://matplotlib.org/stable/users/explain/colors/colormaps.html">https://matplotlib.org/stable/users/explain/colors/colormaps.html</a></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>anndata.AnnData</em>) – The annotated data matrix.</p></li>
<li><p><strong>source_annotation</strong> (<em>str</em>) – The source annotation to use for the relational heatmap.</p></li>
<li><p><strong>target_annotation</strong> (<em>str</em>) – The target annotation to use for the relational heatmap.</p></li>
<li><p><strong>color_map</strong> (<em>str</em>) – The color map to use for the relational heatmap. Default is mint.</p></li>
<li><p><strong>**kwargs</strong> (<em>dict</em><em>, </em><em>optional</em>) – Additional keyword arguments. For example, you can pass font_size=12.0.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p>A dictionary containing:
- “figure” (plotly.graph_objs._figure.Figure):</p>
<blockquote>
<div><p>The generated relational heatmap as a Plotly figure.</p>
</div></blockquote>
<ul class="simple">
<li><dl class="simple">
<dt>”file_name” (str):</dt><dd><p>The name of the file where the relational matrix can be saved.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>”data” (pandas.DataFrame):</dt><dd><p>A relational matrix DataFrame with percentage values.
Rows represent source annotations,
columns represent target annotations,
and an additional “total” column sums
the percentages for each source.</p>
</dd>
</dl>
</li>
</ul>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.visualization.sankey_plot">
<span class="sig-prename descclassname"><span class="pre">spac.visualization.</span></span><span class="sig-name descname"><span class="pre">sankey_plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source_annotation</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target_annotation</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source_color_map</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'tab20'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target_color_map</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'tab20c'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sankey_font</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">12.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/visualization.html#sankey_plot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.visualization.sankey_plot" title="Permalink to this definition"></a></dt>
<dd><p>Generates a Sankey plot from the given AnnData object.
The color map refers to matplotlib color maps, default is tab20 for
source annotation, and tab20c for target annotation.
For more information on colormaps, see:
<a class="reference external" href="https://matplotlib.org/stable/users/explain/colors/colormaps.html">https://matplotlib.org/stable/users/explain/colors/colormaps.html</a></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>anndata.AnnData</em>) – The annotated data matrix.</p></li>
<li><p><strong>source_annotation</strong> (<em>str</em>) – The source annotation to use for the Sankey plot.</p></li>
<li><p><strong>target_annotation</strong> (<em>str</em>) – The target annotation to use for the Sankey plot.</p></li>
<li><p><strong>source_color_map</strong> (<em>str</em>) – The color map to use for the source nodes. Default is tab20.</p></li>
<li><p><strong>target_color_map</strong> (<em>str</em>) – The color map to use for the target nodes. Default is tab20c.</p></li>
<li><p><strong>sankey_font</strong> (<em>float</em><em>, </em><em>optional</em>) – The font size to use for the Sankey plot. Defaults to 12.0.</p></li>
<li><p><strong>prefix</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to prefix the target labels with
the source labels. Defaults to True.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The generated Sankey plot.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>plotly.graph_objs._figure.Figure</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.visualization.spatial_plot">
<span class="sig-prename descclassname"><span class="pre">spac.visualization.</span></span><span class="sig-name descname"><span class="pre">spatial_plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spot_size</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vmin</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-999</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vmax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-999</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">feature</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/visualization.html#spatial_plot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.visualization.spatial_plot" title="Permalink to this definition"></a></dt>
<dd><p>Generate the spatial plot of selected features
:param adata: The AnnData object containing target feature and spatial coordinates.
:type adata: anndata.AnnData
:param spot_size: The size of spot on the spatial plot.
:type spot_size: int
:param alpha: The transparency of spots, range from 0 (invisible) to 1 (solid)
:type alpha: float
:param vmin: The lower limit of the feature value for visualization
:type vmin: float or int
:param vmax: The upper limit of the feature value for visualization
:type vmax: float or int
:param feature: The feature to visualize on the spatial plot.</p>
<blockquote>
<div><p>Default None.</p>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>annotation</strong> (<em>str</em>) – The annotation to visualize in the spatial plot.
Can’t be set with feature, default None.</p></li>
<li><p><strong>layer</strong> (<em>str</em>) – Name of the AnnData object layer that wants to be plotted.
By default adata.raw.X is plotted.</p></li>
<li><p><strong>ax</strong> (<em>matplotlib.axes.Axes</em>) – The matplotlib Axes containing the analysis plots.
The returned ax is the passed ax or new ax created.
Only works if plotting a single component.</p></li>
<li><p><strong>**kwargs</strong> – Arguments to pass to matplotlib.pyplot.scatter()</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>Single or a list of class</strong></p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><cite>~matplotlib.axes.Axes</cite>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.visualization.threshold_heatmap">
<span class="sig-prename descclassname"><span class="pre">spac.visualization.</span></span><span class="sig-name descname"><span class="pre">threshold_heatmap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">feature_cutoffs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">swap_axes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/visualization.html#threshold_heatmap"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.visualization.threshold_heatmap" title="Permalink to this definition"></a></dt>
<dd><p>Creates a heatmap for each feature, categorizing intensities into low,
medium, and high based on provided cutoffs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>anndata.AnnData</em>) – AnnData object containing the feature intensities in .X attribute
or specified layer.</p></li>
<li><p><strong>feature_cutoffs</strong> (<em>dict</em>) – Dictionary with feature names as keys and tuples with two intensity
cutoffs as values.</p></li>
<li><p><strong>annotation</strong> (<em>str</em>) – Column name in .obs DataFrame that contains the annotation
used for grouping.</p></li>
<li><p><strong>layer</strong> (<em>str</em><em>, </em><em>optional</em>) – Layer name in adata.layers to use for intensities.
If None, uses .X attribute.</p></li>
<li><p><strong>swap_axes</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, swaps the axes of the heatmap.</p></li>
<li><p><strong>**kwargs</strong> (<em>keyword arguments</em>) – Additional keyword arguments to pass to scanpy’s heatmap function.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A dictionary contains the axes of figures generated in the scanpy
heatmap function.
Consistent Key: ‘heatmap_ax’
Potential Keys includes: ‘groupby_ax’, ‘dendrogram_ax’, and
‘gene_groups_ax’.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dictionary of <code class="xref py py-class docutils literal notranslate"><span class="pre">Axes</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.visualization.tsne_plot">
<span class="sig-prename descclassname"><span class="pre">spac.visualization.</span></span><span class="sig-name descname"><span class="pre">tsne_plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color_column</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/visualization.html#tsne_plot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.visualization.tsne_plot" title="Permalink to this definition"></a></dt>
<dd><p>Visualize scatter plot in tSNE basis.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>anndata.AnnData</em>) – The AnnData object with t-SNE coordinates precomputed by the ‘tsne’
function and stored in ‘adata.obsm[“X_tsne”]’.</p></li>
<li><p><strong>color_column</strong> (<em>str</em><em>, </em><em>optional</em>) – The name of the column to use for coloring the scatter plot points.</p></li>
<li><p><strong>ax</strong> (<em>matplotlib.axes.Axes</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – A matplotlib axes object to plot on.
If not provided, a new figure and axes will be created.</p></li>
<li><p><strong>**kwargs</strong> – Parameters passed to scanpy.pl.tsne function.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>fig</strong> (<em>matplotlib.figure.Figure</em>) – The created figure for the plot.</p></li>
<li><p><strong>ax</strong> (<em>matplotlib.axes.Axes</em>) – The axes of the tsne plot.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.visualization.visualize_2D_scatter">
<span class="sig-prename descclassname"><span class="pre">spac.visualization.</span></span><span class="sig-name descname"><span class="pre">visualize_2D_scatter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">point_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">theme</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotate_centers</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x_axis_title</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Component</span> <span class="pre">1'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y_axis_title</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Component</span> <span class="pre">2'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_title</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color_representation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/visualization.html#visualize_2D_scatter"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.visualization.visualize_2D_scatter" title="Permalink to this definition"></a></dt>
<dd><p>Visualize 2D data using plt.scatter.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<em>array-like</em>) – Coordinates of the data.</p></li>
<li><p><strong>y</strong> (<em>array-like</em>) – Coordinates of the data.</p></li>
<li><p><strong>labels</strong> (<em>array-like</em><em>, </em><em>optional</em>) – Array of labels for the data points. Can be numerical or categorical.</p></li>
<li><p><strong>point_size</strong> (<em>float</em><em>, </em><em>optional</em>) – Size of the points. If None, it will be automatically determined.</p></li>
<li><p><strong>theme</strong> (<em>str</em><em>, </em><em>optional</em>) – Color theme for the plot. Defaults to ‘viridis’ if theme not
recognized. For a list of supported themes, refer to Matplotlib’s
colormap documentation:
<a class="reference external" href="https://matplotlib.org/stable/tutorials/colors/colormaps.html">https://matplotlib.org/stable/tutorials/colors/colormaps.html</a></p></li>
<li><p><strong>ax</strong> (<em>matplotlib.axes.Axes</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – Matplotlib axis object. If None, a new one is created.</p></li>
<li><p><strong>annotate_centers</strong> (<em>bool</em><em>, </em><em>optional</em><em> (</em><em>default: False</em><em>)</em>) – Annotate the centers of clusters if labels are categorical.</p></li>
<li><p><strong>x_axis_title</strong> (<em>str</em><em>, </em><em>optional</em>) – Title for the x-axis.</p></li>
<li><p><strong>y_axis_title</strong> (<em>str</em><em>, </em><em>optional</em>) – Title for the y-axis.</p></li>
<li><p><strong>plot_title</strong> (<em>str</em><em>, </em><em>optional</em>) – Title for the plot.</p></li>
<li><p><strong>color_representation</strong> (<em>str</em><em>, </em><em>optional</em>) – Description of what the colors represent.</p></li>
<li><p><strong>**kwargs</strong> – Additional keyword arguments passed to plt.scatter.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>fig</strong> (<em>matplotlib.figure.Figure</em>) – The figure of the plot.</p></li>
<li><p><strong>ax</strong> (<em>matplotlib.axes.Axes</em>) – The axes of the plot.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spac.visualization.visualize_nearest_neighbor">
<span class="sig-prename descclassname"><span class="pre">spac.visualization.</span></span><span class="sig-name descname"><span class="pre">visualize_nearest_neighbor</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">distance_from</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">distance_to</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stratify_by</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spatial_distance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'spatial_distance'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">facet_plot</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation_colorscale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'rainbow'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">defined_color_map</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/spac/visualization.html#visualize_nearest_neighbor"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spac.visualization.visualize_nearest_neighbor" title="Permalink to this definition"></a></dt>
<dd><p>Visualize nearest-neighbor (spatial distance) data between groups of cells
with optional pin-color map via numeric or distribution plots.</p>
<p>This landing function first constructs a tidy long-form DataFrame via
function <cite>_prepare_spatial_distance_data</cite>, then dispatches plotting to
function <cite>_plot_spatial_distance_dispatch</cite>.  A pin-color feature guarantees
consistent mapping from annotation labels to colors across figures,
drawing the mapping from <code class="docutils literal notranslate"><span class="pre">adata.uns</span></code> (if present) or generating one
automatically through <cite>spac.utils.color_mapping</cite>.</p>
<dl class="simple">
<dt>Plot arrangement logic:</dt><dd><ol class="arabic simple">
<li><p>If stratify_by is not None and facet_plot=True =&gt; single figure
with subplots (faceted).</p></li>
<li><p>If stratify_by is not None and facet_plot=False =&gt; multiple separate
figures, one per group.</p></li>
<li><p>If stratify_by is None =&gt; a single figure with one plot.</p></li>
</ol>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>anndata.AnnData</em>) – Annotated data matrix with distances in <cite>adata.obsm[spatial_distance]</cite>.</p></li>
<li><p><strong>annotation</strong> (<em>str</em>) – Column in <cite>adata.obs</cite> containing cell phenotypes or annotations.</p></li>
<li><p><strong>distance_from</strong> (<em>str</em>) – Reference phenotype from which distances are measured. Required.</p></li>
<li><p><strong>distance_to</strong> (<em>str</em><em> or </em><em>list</em><em> of </em><em>str</em><em>, </em><em>optional</em>) – Target phenotype(s) to measure distance to. If None, uses all
available phenotypes.</p></li>
<li><p><strong>stratify_by</strong> (<em>str</em><em>, </em><em>optional</em>) – Column in <cite>adata.obs</cite> used to group or stratify data (e.g. imageid).</p></li>
<li><p><strong>spatial_distance</strong> (<em>str</em><em>, </em><em>optional</em>) – Key in <cite>adata.obsm</cite> storing the distance DataFrame. Default is
‘spatial_distance’.</p></li>
<li><p><strong>facet_plot</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True (and stratify_by is not None), subplots in a single figure.
Otherwise, multiple or single figure(s).</p></li>
<li><p><strong>method</strong> (<em>{'numeric'</em><em>, </em><em>'distribution'}</em>) – Determines the plotting style (catplot vs displot).</p></li>
<li><p><strong>plot_type</strong> (<em>str</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – Specific seaborn plot kind.  If None, sensible defaults are selected
(‘boxen’ for numeric, ‘violin’ for distribution).
For method=’numeric’: ‘box’, ‘violin’, ‘boxen’, ‘strip’, ‘swarm’.
For method=’distribution’: ‘hist’, ‘kde’, ‘ecdf’.</p></li>
<li><p><strong>log</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, applies np.log1p transform to the distance values.</p></li>
<li><p><strong>annotation_colorscale</strong> (<em>str</em><em>, </em><em>optional</em>) – Matplotlib colormap name used when auto-enerating a new mapping.
Ignored if ‘defined_color_map’ is provided.</p></li>
<li><p><strong>defined_color_map</strong> (<em>str</em><em>, </em><em>optional</em>) – Key in ‘adata.uns’ holding a pre-computed color dictionary.
Falls back to automatic generation from ‘annotation’ values.</p></li>
<li><p><strong>ax</strong> (<em>matplotlib.axes.Axes</em><em>, </em><em>optional</em>) – The matplotlib Axes containing the analysis plots.
The returned ax is the passed ax or new ax created.
Only works if plotting a single component. Default is None.</p></li>
<li><p><strong>**kwargs</strong> (<em>dict</em>) – Additional arguments for seaborn figure-level functions.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>{</dt><dd><p>‘data’: pd.DataFrame,  # long-form table for plotting
‘fig’ : matplotlib.figure.Figure | list[Figure] | None,
‘ax’: matplotlib.axes.Axes | list[matplotlib.axes.Axes],
‘palette’: dict  # {label: ‘#rrggbb’}</p>
</dd>
</dl>
<p>}</p>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If required parameters are invalid.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">res</span> <span class="o">=</span> <span class="n">visualize_nearest_neighbor</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">adata</span><span class="o">=</span><span class="n">my_adata</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">annotation</span><span class="o">=</span><span class="s1">&#39;cell_type&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">distance_from</span><span class="o">=</span><span class="s1">&#39;Tumour&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">distance_to</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Stroma&#39;</span><span class="p">,</span> <span class="s1">&#39;B cell&#39;</span><span class="p">],</span>
<span class="gp">... </span>    <span class="n">method</span><span class="o">=</span><span class="s1">&#39;numeric&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">plot_type</span><span class="o">=</span><span class="s1">&#39;box&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">facet_plot</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">stratify_by</span><span class="o">=</span><span class="s1">&#39;image_id&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">defined_color_map</span><span class="o">=</span><span class="s1">&#39;pin_color_map&#39;</span>
<span class="gp">... </span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fig</span> <span class="o">=</span> <span class="n">res</span><span class="p">[</span><span class="s1">&#39;fig&#39;</span><span class="p">]</span>      <span class="c1"># matplotlib.figure.Figure</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ax_list</span> <span class="o">=</span> <span class="n">res</span><span class="p">[</span><span class="s1">&#39;ax&#39;</span><span class="p">]</span>   <span class="c1"># list[matplotlib.axes.Axes] (faceted plot)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span>  <span class="o">=</span> <span class="n">res</span><span class="p">[</span><span class="s1">&#39;data&#39;</span><span class="p">]</span>     <span class="c1"># long-form DataFrame</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ax_list</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Tumour → Stroma distances&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

</section>


           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2025, Fang Liu, Rui He, and George Zaki.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>